Enhancements to Globalnet Implementation (#22)
submariner-io/enhancements: submariner/globalnet-enhancement2-0.md

the cluster where the source IP of the Pod is modified to the corresponding globalIp. This
translation enables us to uniquely identify the traffic from the Pod, when it reaches the
destination cluster.

Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
node of the cluster, where the service resides. The ingress rules enable any incoming traffic
destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.

In order to support connectivity from the Host Network to the remote services, Globalnet also
annotates every node in the cluster with a unique globalIp.

While this works fine for small clusters, where the number of Pods and Services are limited, for
large clusters it poses scalability and performance issues as we would eventually run out of
globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
connectivity as well as Headless Services are not supported.

## Proposal

This enhancement proposes the following changes to make Globalnet implementation scale better.

> 1. By default, every cluster will be assigned a single globalIp which will be used as egress IP

Reviewer: This seems like we could easily reach a scale limit due to limited number of ports available.
I suggest we allow either IP or CIDR here to allow good scalability for bigger deployments.

Reviewer 2: That could be achieved by adding per-namespace IPs or selector based IPs, but it's a good point, may be we can reserve a CIDR for that and split the IPs based on the internal cluster CIDRs?

Author: That's a good point @mkolesnik. For a single SNAT EgressIP, depending on the configured ephemeral ports, the max number of connections that can be supported will be 65K. But, as you know, this is the active connections at any point of time. Globalnet (or for that matter Submariner) does not maintain the connection info and its handled by Linux Conntrack, so these rules have to be programmed ahead of time. 

IPTables supports providing a contiguous range (not CIDR) while configuring an SNAT rule, so we can specify couple of IPs as globalIPs at ClusterLevel/NamespaceLevel. Have to check how to model this...

Example: iptables -t nat -I SUBMARINER-GN-EGRESS -p all -s 10.0.128.11 -j SNAT --to 169.254.2.11-169.254.2.14

Author: For the initial implementation, I feel we can allocate two globalIPs at the cluster level. In future we can make it configurable entity. Your thoughts?

   for any cross-cluster communication.

> 2. Provision to allocate a globalIp at the namespace/project level which takes precedence over

Reviewer: Same here, we should allow a CIDR (or single IP).

Reviewer 2: it's a good point

Author: AFAICT a single globalIP could be sufficient at the namespace level. 

Let me explain:
In Linux, ephemeral port range can be seen/configured via `sysctl net.ipv4.ip_local_port_range`. The default range is `net.ipv4.ip_local_port_range = 32768    60999` (i.e., approx 27K). Connection tracking is done based on 5 tuple - i.e., SrcIP, SrcPort, DestIP, DestPort and Protocol. In our use-case we are talking about Egress traffic at namespace level, so SrcIP is globalIP allocated to namespace and SrcPort is ephemeral port range.

In order for ephemeral port exhaustion to happen at the namespace level, if all the Pods in the namespace connect to the same DestIP:DestPort we will run out after 27K connections are established. 

Please note that if pods try to connect to different DestIPs and/or DestPorts, each combination has 27K limit.
So based on this, IMHO a single globalIP should suffice. 

If you think, I'm wrong or missing something, do let me know.

@mkolesnik @mangelajo @aswinsuryan @nyechiel

   the globalIp allocated to the cluster.

> 3. Provision to allocate a globalIp to a Pod (or a set of Pods) which has higher precedence

Reviewer: I feel like at the pod level (and probably this can apply to namespace level as well) explicit is better than implicit, so it seems to me we can have the "global ip policy" annotate the pods with the global IP.

This will give the user visibility which global IP is used for which pod which can be great for debugging.

This is only for pods expliditly selected, if pod is not selected it will fall back to namespace/cluster policy.

I think it'll also make the whole thing very easy for us to compute.

EG when defining rules on the gateway we can put pod rules first, then catch all rules for namespaces, and finally the rule for the cluster.

Reviewer 2: yup, it could be done in terms of iptables with chains, you have a chain for pods, a chain for namespaces, a general chain as a fallback, @mkimuram suggested this.

Author: Yes, this is possible and we already do something similar in the current Globalnet implementation. We first send packets to SUBMARINER-GN-MARK chain to mark packets destined to remoteClusters and in the subsequent chains (which are like fallback chains) we use the mark to make decisions. Any traffic which does not include the mark (for example external traffic) does not hit our rules.

>    over globalIPs at namespace level. If there are multiple policies, users can specify a priority

Reviewer: Priority is not really an important feature IMHO but for sure we want consistency, so that a single pod doesnt flip between IPs.
I feel how we do it is better left as implementation details (as user will likely not care).

Reviewer 2: It's something that we have identified in brainstorms with the community.

Reviewer 2: But may be it's worth leaving this for the future, as it's a simple additive improvement, and simplifies the iptables a lot. (Also probably in OVN)

Author: Say we remove Priority (atleast for the initial implementation) and user creates multiple GlobalnetEgressIP CRDs where PodSelectors match common Pods. The behavior (ie., globalIP which will be used) cannot be well defined. Would this be acceptable?

CC: @nyechiel

Reviewer 2: @sridhargaddam it could be that one policy matches a namespace, and another policy specific pods.

May be we should put those in separate iptable chains, giving namespace only selectors a first priority chain, and more complex selectors a lower priority.

Then in the future we can think about priority. I suspect this would satisfy most cases.

Reviewer: As I specified in my now outdated comment, I feel like at the pod level (and probably this can apply to namespace level as well) explicit is better than implicit, so it seems to me we can have the "global ip policy" annotate the pods with the global IP.

This will prevent any "global IP flipping" as feared here, and will give better visibility to the users to which IP is used (in case they want to of course).

Author: Okay, what about multiple GlobalnetEgressIP objects where PodSelectors match common Pods.

i.e., say we have pods - p1, p2, p3, p4 and p5.
GlobalnetEgressIP object1 - matches p1 and p2
GlobalnetEgressIP object2 - matches p2 and p4

What should be the egress globalIP for p2 which has a match in both the objects? Can we say that it will be implementation-specific?

Reviewer 2: We should say it's not deterministic, and any of both could be expected.

Author: Thanks @mangelajo. I had look at OpenShift EgressIP feature and even over there its mentioned that its non deterministic.

"_When a pod matches the selector for multiple EgressIP objects, there is no guarantee which of the egress IP addresses that are specified in the EgressIP objects is assigned as the egress IP address for the pod._"

Author: I'm removing the Priority field.

   field for each of the policy.
4. Instead of annotating every ClusterIP Service in the cluster, only exported Services will be

>    annotated with a globalIp.

Reviewer: This is great and I feel like we can start doing it regardless of the other proposals straight away, kind of "globalnet 1.5"

Author: I dont know if we can backport this to 0.8.2 (or something) as its a feature and user-experience will be different. But this can surely go as part of the initial 0.9 release if other Globalnet 2.0 changes take some additional time.

Reviewer: Yeah intention wasn't to backport but rather as part of 0.9

Reviewer 2: I think it's better not to spend more time in the existing globalnet, and go ahead using our time with v2

Reviewer 3: This is already in and is rather independent of everything else we're doing in v2. Maybe can remove this  from this spec?

Author: But since we will not be backporting it to 0.8 release. Lets keep it part of Globalnet 2.0 work.

> 5. For Headless Services, necessary ingress rules will be programmed to support connectivity to

Reviewer 2: IMHO egress IP should be kept independent of ingress IP.

For example, for the external network use cases, where people want a consistent IP for pods (same IP for a namespace or set of pods, or cluster...) ... that will make varying IPS for the pods as soon as they restart.

So for that I propose that this would only talk of the ingress IP.

Author: > IMHO egress IP should be kept independent of ingress IP.

For a Headless Service backed by StatefulSet, AFAIU even K8s does not guarantee the same PodIP when a Pod is restarted. It is recommended that clients use DNS to obtain the IPaddress of the Pods.
When DNS returns the SRV records, it points to the record entries that contain the Pod IPs. We are doing the same exact thing in the architecture and returning the globalIPs (of the Pods) which can be used for both ingress/egress. We are more aligned to K8s in this architecture of Globalnet.

> For example, for the external network use cases, where people want a consistent IP for pods (same IP for a namespace or set of pods, or cluster...) ... that will make varying IPS for the pods as soon as they restart.

As you know, if users want a consistent IP, they should use a Service infront of the Pod which is the recommended approach.

Reviewer 2: We are trying to satisfy the 2nd requirement that @mkimuram found by separating those details in the design.

This is important for them because they use source IP for authentication (typical of mysql  for example.)

Reviewer 2: And I think it's natural, I don't believe the egress IP of a POD needs to be modified by a service definition,

Reviewer 2: What if... 

1) When there is no specific GlobalnetEgressIP, the egress IP will be the one of the pod in the headless services
2) If there is one GlobalnetEgressIP requested, that will take precedence?

That way we can satisfy both cases?

It would take precedence over the cluster global IP :)

I make harder design, but I'm trying to satisfy all possible presented use cases.

Author: AFAIU the 2nd requirement from @mkimuram is

When a Pod (aka Client) tries to talk to some External application, the sourceIP seen by External application should not change even when the Pod restarts. 
This will allow External application to use SG/ACLs to allow traffic from a particular IP.

To achive this use-case, in the new design we have the following possibilities

1. User can explicitly request for a GlobalnetEgressIP at the namespace level. But the drawback is that all the Pods in that namespace will have access to the External application.

2. User can explicitly request for a GlobalnetEgressIP for the particular Pod using PodSelectors. Assuming that there is only a single client Pod, user can even specify the globalIP that he wants the Pod to use.
   This will enable Globalnet to assign the same globalIP to the Pod even when the Pod restarts.
   
3. User creates headless service with backend Pods and always uses DNS/LH to resolve the latest IPs and configure appropriate ACL/SG rules on the external application.

Reviewer 2: In my awful diagram that we shared over bluejeans:

```

GlobalnetEgressIP:
   PodSelector:
      app=http1
                
   Status:
      GlobalIP: 169.254.0.11
               
               
  http-1234  10.0.0.1     <------  169.254.0.12
  http-3333  10.0.0.2     <------  169.254.0.13    -- 169.254.0.11 ----->      ExternalNetwork ------->       10.1.0.3   (database-A)   ok, so PWD OK, IP==169.254.0.11 --> login ok
  http-4444  10.0.0.3     <------  169.254.0.14                                                                                         ok, so PWD OK, IP!=169.254.0.11 --> login FAILED
  
  
          Service:
            selector:
               app=http1
            ClusterIP: None
            
            
   If there is no GlobalnetEgressIP selector on those pods
  
                 
  http-1234  10.0.0.1     <------  169.254.0.12    -- 169.254.0.12 ----->
  http-3333  10.0.0.2     <------  169.254.0.13    -- 169.254.0.13 ----->    ExternalNetwork/Other cluster ------->       10.1.0.3  
  http-4444  10.0.0.3     <------  169.254.0.14    -- 169.254.0.14 ----->                                                                                                      
  
  
  
 
```

Author: Done

   the backend Pods. In a Headless Service use-case, every backend Pod will be annotated with
   a unique globalIP and the same globalIP will be used for both Ingress as-well-as Egress.
   globalIPs allocated for Pods backed by Headless Services will have highest priority over other
   egress policies.

![GlobalnetTopology](./images/globalnet-topology.png)

If the external application in a remote Cluster is interested in consuming the exported
Service, it will have to reach the globalIP assigned to the exportedService and the response
from the backend Pods will have the source as the globalIP of the exportedService.

On the other hand, when the backend Pod tries to talk to the external application in remote
Cluster, the external application will see that traffic is coming with a globalIP that is
assigned either at the Cluster level, or namespace level, or the one requested by the user
at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
for both ingress and egress traffic can create Headless Services in front of them.

> The proposal has its own set of Pros and Cons.

Author: Currently, while deploying Submariner Globalnet, we can provide an env variable (SUBMARINER_EXCLUDE_NS) to inform Globalnet to Skip annotating any Pods/Services in that Namespace. 

In the new design, do we want to continue to support it or deprecate it given that we are now using CRDs to annotate at Namespace/Pod level. Any suggestions?

Author: Related: Based on the discussion with @mangelajo, to support Globalnet2.0 we will have to break the backward compatibility as we might be using two annotations (one for Egress and the other for Ingress).

Reviewer 2: We can deprecate that. I think it could be useful to let users mark specific system namespaces sometimes, I see use cases for that. (like connecting to a remote API, etc..)

Author: Thanks @mangelajo. So, we'll deprecate SUBMARINER_EXCLUDE_NS.

Regarding, "I think it could be useful to let users mark specific system namespaces sometimes" -> For this, are you suggesting that we support a new Env variable that instructs Globalnet to automatically annotate all Services in those namespaces? If so, in most of the common use-cases we may not really require accessing those Services and even if there is a need, user can explicitly export those services right which will keep the behavior consistent? WDYS?

Reviewer 3: +1 to deprecate `SUBMARINER_EXCLUDE_NS`.

> I think it could be useful to let users mark specific system namespaces sometimes

Should not be an ENV variable. This is something that was briefly discussed in MCS API. We can have an annotation or label added to a namespace, say `NamespaceExport`, and then we auto-export all services in that namespace. Export of service is what will lead to globalIP for it. This retains design approach of requiring to export services, but providing a sort of mechanism to export services in a namespace.

Author: Good point. I'd prefer that we do that as a future enhancement (should not be too difficult once the basic implementation is in place). Currently, AFAICT we will have a lot of work from an implementation POV to support what is specified in this enhancement.

and the corresponding egress rules are programmed on the node, it was not able to communicate with
remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
have a globalIp at the cluster level, a Pod can communicate with remote Services right away which
will be closer to the Vanilla Submariner experience.

## Cons

A limitation with the proposed approach is that we will lose the unique identity for **every**
Pod which is supported today. For users that intend to use Network Policies across Clusters or
for debugging purposes, this can be mitigated by creating the necessary CRDs which create unique
globalIps at the Namespace/Pod level.

The following CRD is proposed to support the new design.

```Go
type GlobalnetEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec               EgressIPSpec `json:"spec"`
}

type EgressIPSpec struct {
    // Selects the pods to which the GlobalnetEgressIP will be applied.
    // If multiple pods match the selector the same egressIP will be used for all those pods.
    // +optional

>     PodSelector PodSelectorSpec `json:"podSelector,omitempty"`

Reviewer 2: Let's use the standard kubernetes selectors instead:
```
k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelector
```
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L1081

This will allow use to use MatchExpressions for free... the implementation of the matching is given to us.

Example of matching:
https://github.com/submariner-io/coastguard/blob/devel/pkg/networkpolicy/networkpolicy.go#L147

Author: Yes, I saw this but I intentionally considered the PodSelector. 

IMHO, I think we don't really need the complex LabelSelector for our use-case. Network Policies might have their reasons, but GlobalnetEgressIP CRD is more like a Service that is backed by Pods. Service CRD also uses PodSelector. What do you say?

Reviewer 4: [This discussion](https://github.com/kubernetes/kubernetes/issues/48528) is related? I think that most of the use cases would be covered by the same interface to Service, but some might find useful to the "complex" selector, as there seems such a use case even with Service.

Reviewer 2: The LabelSelector brings greater flexibility to the user, without (IMO at least) greater cost of development, or worsened UI.

The fact that ServiceSelector uses just a map lot labels I believe it has to do with it being much older (and more core) than network policies, which makes it harder to upgrade. In this case it's a fresh start.

Author: Done

    // Clients can request a specific globalIP to be allocated. If the requested
    // globalIP is part of globalCIDR of the Cluster and is not allocated to other
    // K8s objects, then it will be allocated, otherwise creation of the CRD will fail.
    // Valid values are empty string (""), or a valid IP address from the globalCIDR of
    // the Cluster. This field cannot be changed through updates.
    GlobalIP    string          `json:"private_ip"`
}

type PodSelectorSpec struct {
    // MatchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
    // map is equivalent to an element of matchExpressions, whose key field is "key", the
    // operator is "In", and the values array contains only "value". The requirements are ANDed.
    MatchLabels map[string]string `json:"matchLabels"`

    // Associated priority to be used for the selected pods.
    // Its value should be between 0 (lowest) and 1000 (highest) inclusive.
    // When this value is omitted, the default value of 0 is used.
    // +optional

>     Priority    int32 `json:"priority,omitempty"`

Reviewer 2: Because of the LabelSelectors change, let's move this priority up to the EgressIPSpec

Author: As you know, IPTables do not have priority fields like openflow rules. So care has to be taken to program rules in the right order to ensure priority is retained. Also using separate Chains (like a chain name for ClusterLevelGlobalIP, NameSpaceLevelGlobalIP and a chain for PodLevelGlobalIPs) will make debugging/investigating issues a bit easier. 

So when users explicitly require IPs for Pods, they will carry higher priority over namespace. So, priority is a field that goes along with PodSelector.

Reviewer 2: But won't that order of iptables will be the same regardless of where we put the priority field (here or in the GlobalnetEgressIP) ? may be I'm missing something

Author: @mangelajo From an implementation POV, my plan is to create Separate Chains for each level.
For example: SUBMARINER-GN-PODS, SUBMARINER-GN-NAMESPACES and SUBMARINER-GN-CLUSTER

So, for egress traffic destined to remoteClusters we will have to use a combination of marking and ipSets to redirect traffic as follows.

```
+--------------------+
|                    |
|  Egress Traffic    +----+
|                    |    |
+--------------------+    |
                          |
                          v
         +----------------+---+
         |                    |
         | SUBMARINER-GN-PODS +-------+
         |                    |       |
         +--------------------+       |
                                      |
                                      v
                        +-------------+---------+
                        |                       |
                        | SUBMARINER-GN-NAMESPACE
                        |                       |
                        +---------------------+-+
                                              |
                                              v
                                +-------------+-------+
                                |                     |
                                | SUBMARINER-GN-CLUSTER
                                |                     |
                                +---------------------+
```

The order of iptables (based on the priority) will matter mostly in the `SUBMARINER-GN-PODS` chain. If we put the priority in the `Spec` as shown below, it can imply that globalIP associated with a namespace based GlobalnetEgressIP object (which does not have any podSelectors) can have higher priority over GlobalnetEgressIP that uses PodSelectors. 

i.e., 
```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: ns-egressip
     namespace: myproject
   spec:
     priority: 100            <-- This will be confusing to users.
```

Happy to explain over a call if required.

Reviewer 2: Ahh, I see what you mean. But you could still create an empty podselector, with just priority. I see what you mean though.

Reviewer 2: I say let's drop the priority altogether, then we can put the PodSelector on the top level of the spec.

Reviewer 3: Why no `NamespaceSelector` For the use case to allocate GlobalIP to namespace, won't it be better to create a GlobalnetEgressIP object with NamespaceSelector? I realize that scope namespace with empty podSelector can achieve that, but thinking from an ownership perspective. Who decides which namespaces to get their own globalIP, user owning the namespace or admin owning the cluster? Thinking about RBAC implications of the two approaches.

Author: Done

2.In a Globalnet deployment user wants the Pods in a particular namespace to use a unique globalIP
  instead of the globalIP assigned at the cluster level.

  To achieve this, one can apply the following CRD.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: ns-egressip
     namespace: myproject
   spec:
     globalIP: (optional)
```

3.In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace
  to use a unique globalIP as egressIP for cross-cluster communication.

  To achieve this, one can apply the following CRD.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: db-pods
     namespace: myproject
   spec:
     globalIP: (optional)
     podSelector:
       matchLabels:
         role: db
       priority: 100  
```

4.In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace
  to use a unique globalIP both as ingress and egressIP for cross-cluster communication.

  To achieve this, user can create a Headless Service (backed by a Deployment or StatefulSets) and

>   Globalnet will ensure that all the backend Pods that belong to the Headless Service are assigned

Reviewer 4: Apparently, it doesn't seem to allow users to specify a particular globalIP to a selected Pod (or set of Pods) via `GlobalnetEgressIP` for this case. Can it be specified? Also, does it allow the globalIP not to be changed?

Reviewer 2: Yeah, see the comments in here: https://github.com/submariner-io/enhancements/pull/22/files#r572131272

I'm suggesting that order of priorities to allow what you propose.

Author: @mkimuram, as @mangelajo mentioned, we will be updating the priorities for GlobalnetEgressIP (to take highest precedence) and this will enable us to use the same globalIP for egress even when Pods restart. I'll update this PR shortly.

translation enables us to uniquely identify the traffic from the Pod, when it reaches the
destination cluster.

Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
node of the cluster, where the service resides. The ingress rules enable any incoming traffic
destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.

In order to support connectivity from the Host Network to the remote services, Globalnet also
annotates every node in the cluster with a unique globalIp.

While this works fine for small clusters, where the number of Pods and Services are limited, for
large clusters it poses scalability and performance issues as we would eventually run out of
globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
connectivity as well as Headless Services are not supported.

## Proposal

This enhancement proposes the following changes to make Globalnet implementation scale better.

Egress traffic:

> 1. By default, every cluster will be assigned two globalIps which will be used as egress IPs

Reviewer 3: Why limit to 2? Why not make it a configurable that can be scaled up/down with default of 2 or 1?

Author: Okay done.

  Service Pods as shown below.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: http-pods
     namespace: myproject
   spec:
     globalIP: (optional)
     selector:
       podSelector:
         matchLabels:
           app: http
   status:
     globalIP: 169.254.0.100
```

![GlobalnetTopology](./images/globalnet-svc.png)

## Alternatives

It is possible to support Pod to Pod communication even with Globalnet. However, to support this, it
requires programming additional IPtable rules on the nodes which will have an impact on the performance
of the solution. This proposal provides some limited Pod to Pod connectivity via the Headless Services.
If user is interested in Pod to Pod communication for all the Pods, it is recommended to install
Clusters with non-overlapping CIDRs and use Vanilla Submariner instead of Globalnet.

## Work items

1. Support annotation at the cluster level and program necessary egress rules.
2. Support annotation at namespace level which takes precedence over globalIp at cluster level.
3. Support annotation to Pods (or set of Pods) which has the highest precedence while
   honouring the priority.

> 4. Annotate only exported Services instead of annotating all Services.

Reviewer 3: As mentioned earlier, since this is already done, can consider removing it.

Author: Lets not remove it, we are not backporting this feature to current GN release.

The following CRD is proposed to support the new design.

```Go
type GlobalnetEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
    Spec               EgressIPSpec `json:"spec"`

    // Observed status of GlobalnetEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type EgressIPSpec struct {
    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
    // K8s objects, then it will be allocated and the status field will be updated accordingly.
    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
    // If this field is not specified, a single globalIP will be allocated.
    // This field cannot be changed through updates.
    // +optional
    GlobalIPs  []string `json:"globalIPs,omitempty"`

    // Selects the Namespaces to which this GlobalnetEgressIP object applies.

>     // This field is mandatory.

Reviewer 2: Let's make this one not mandatory, so it would work like the network policies, where you can select only based on the podselector if you want.

assigned either at the Cluster level, or namespace level, or the one requested by the user
at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
for both ingress and egress traffic can create Headless Services in front of them and should
avoid requesting explicit globalIP that match Headless Service backend Pods.

The proposal has its own set of Pros and Cons.

## Pros

Since we will be annotating only select K8s objects with globalIP, it will help in improved scalability.
Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
improve the overall performance of the solution.

In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
and the corresponding egress rules are programmed on the node, it was not able to communicate with
remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
have a globalIP at the cluster level, a Pod can communicate with remote Services right away which

> will be closer to the Vanilla Submariner experience.

Reviewer: This could still have race conditions AFAICT if traffic first exits with the Cluster/Namespace IP and later with a Pod IP.
If I'm wrong, cool, but if my assessment is correct I would list this as a caveat (since the whole architecture is asynchronous).

Author: You are right @mkolesnik. I'll put this under Caveat section.

## Cons

A limitation with the proposed approach is that we will lose the unique identity for **every**
Pod which is supported today. For users that intend to use Network Policies across Clusters or
for debugging purposes, this can be mitigated by creating the necessary GlobalnetEgressIP CRDs
which create unique globalIPs at the Namespace/Pod level.

> The following CRD is proposed to support the new design.

Reviewer: Shouldn't this be in a section?
Right now it'll be under "Cons"

Author: Done

    // In case it is set, its intersected with NamespaceSelector, and only Pods that match the
    // PodSelector and present in the NamespaceSelector will have the GlobalIPs as EgressIPs.
    // Either namespaceSelector or podSelector have to be specified. If both are omitted,
    // its considered as an error.
    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

type EgressIPStatus struct {
    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
    GlobalIPs []string `json:"globalIPs"`

    // The Namespaces to which the GlobalIPs are applied.
    Namespaces []string `json:"namespaces"`

    // The Pods to which the GlobalIPs are applied.
    Pods []string `json:"pods"`
}
```

Creation of GlobalnetEgressIP object will be controlled via a ClusterRole and only users having
the necessary role will be able to create the objects. The GlobalnetEgressIP object can be
created in any namespace with selectors that match any other namespaces.

> ### User Stories

Reviewer: Would older globalnet annotation still be supported? It's not clear

Author: Done.

Reviewer: Can you provide examples for ingress IPs?

Author: The GlobalnetEgressIP CRD is for EgressIP only. 
By default, all exported Services (including Headless Services) will automatically get a globalIP (annotated on the Service object or ServiceExport object) and Globalnet will program rules to support **Ingress** traffic destined to the corresponding globalIP. 

Do you want me to add a use-case to show how/when Ingress globalIP is allocated to an exported service as shown below.

```
kubectl -n default create deployment nginx --image=nginxinc/nginx-unprivileged:stable-alpine
kubectl -n default expose deployment nginx --port=8080
subctl export service --namespace default nginx
```

Reviewer: Yes I think it'll clarify it.
Also it should be clear that the only way to get an ingress IP is by exporting a service (clusterip of headless) and that users can't assign specific IPs.

Author: I thought it was clear from the name GlobalnetEgressIP :) I'll add it.

have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
will be closer to the Vanilla Submariner experience.

## Design Details

The following CRD is proposed to support the new design.

```Go
type GlobalnetEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
    Spec               EgressIPSpec `json:"spec"`

    // Observed status of GlobalnetEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type EgressIPSpec struct {
    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
    // K8s objects, then it will be allocated and the status field will be updated accordingly.
    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
    // If this field is not specified, a single globalIP will be allocated.
    // This field cannot be changed through updates.
    // +optional

>     GlobalIPs  []string `json:"globalIPs,omitempty"`

Reviewer 2: Can we do this?, or should we have GlobalIPs CIDR?

Reviewer 2: What if we have 

     GlobalIPCIDR string ...    (optional)

Which could be used like
 ```
     GlobalIPCIDR: 169.254.4.1/32  or 169.254.4.1
     GlobalIPCIDR: 169.254.4.0/29
```
or
```
    GlobalIPCIDR: "/29"   meaning that you want a /29 and you don't care about the 
```

Author: IMHO, let's keep it simple. The more options we provide the more corner/edge cases that needs to be handled. 

Since these are EgressIPs, from an implementation POV, if we take CIDR input, we still have to specify individual IPTable SNAT rules (AFAICT IPTables does not support specifying a CIDR for SNAT target) when we are programming rules on the node.

On a related note, OpenShift EgressIP feature does not support CIDRs, it only supports an array. I'm trying to keep our GlobalnetEgressIP feature a bit aligned to the EgressIP use-case from OCP.

In most of the cases, users will just need a single GlobalIP. The array is to provide them flexibility to specify multiple GlobalIPs. If we really see a need for CIDRs, we can do it as future enhancement once the implementation matures.

Reviewer: While this seems logical, the main problem with using a list is the ability to specify non continuous IPs. This is not supported by iptables` SNAT action.

If you don't want to have CIDR here, then we should either have an IP and an optional endIP field, or validation to make sure the array is continuous. Or we can make sure the array size is 1 or 2..

I find having 2 distinct fields easier in this regard.

```
       --to-source [ipaddr[-ipaddr]][:port[-port]]
              which can specify a single new source IP address, an inclusive range of IP addresses. Optionally a port  range,  if  the  rule  also
              specifies  one  of the following protocols: tcp, udp, dccp or sctp.  If no port range is specified, then source ports below 512 will
              be mapped to other ports below 512: those between 512 and 1023 inclusive will be mapped to ports below 1024, and other ports will be
              mapped  to  1024  or above. Where possible, no port alteration will occur.  In Kernels up to 2.6.10, you can add several --to-source
              options. For those kernels, if you specify more than one source address,  either  via  an  address  range  or  multiple  --to-source
              options, a simple round-robin (one after another in cycle) takes place between these addresses.  Later Kernels (>= 2.6.11-rc1) don't
              have the ability to NAT to multiple ranges anymore.
```

Author: Two distinct fields would be easier I guess. What is your opinion on the following 

```
// User can explicitly request a specific GlobalIP and this should be a
// valid IPaddress from globalnetCIDR allocated to the Cluster. 
// If the requested globalIP falls outside of the globalnetCIDR or is
// already allocated, then Status field will be set to Error with
// appropriate message.
// If this field is not specified, a single globalIP will be allocated.
// This field cannot be changed through updates.
// +optional
GlobalIP      string  `json:"globalIP,omitempty"`

// The number of globalIPs requested.
// User can specify this field while omitting GlobalIP. In such cases,
// Globalnet will allocate the requested number of contiguous GlobalIPs
// for this GlobalnetEgressIP object.
// If unspecified, NumGlobalIPs defaults to 1.
// +optional
NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
```

Reviewer: Yeah it looks good.

Not sure why the fields shouldn't be able to change though? Perhaps initially it makes sense, but generally what's wrong with changing the global IP?

Author: Technically we can support it. But in the initial design, to keep it simple I want to avoid supporting updating the globalIP field as we will have to update the associated IPTable rules on the GatewayNode (and any associated annotations on the objects). We can support it in the future once the basic implementation is mature/stable.

Reviewer: OK sure just wanted to make sure I'm understanding correctly. It sounded like a prioritization issue to me, rather than a real limitation.

    // On the other hand, if you omit specifying namespaceSelector, it does not select any
    // namespaces and selects only Pods from the namespace where the GlobalnetEgressIP is
    // deployed to.
    // +optional
    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`

    // Selects the pods whose label match the definition. This is an optional field and
    // in case it is not set, it results in all the Pods selected from the NamespaceSelector.
    // In case it is set, its intersected with NamespaceSelector, and only Pods that match the
    // PodSelector and present in the NamespaceSelector will have the GlobalIPs as EgressIPs.
    // Either namespaceSelector or podSelector have to be specified. If both are omitted,
    // its considered as an error.
    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

type EgressIPStatus struct {
    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.

>     GlobalIPs []string `json:"globalIPs"`

Reviewer 2: Same about the list

Reviewer: Is this list expected to be different than the one in spec?

Author: Spec.GlobalIPs is an optional field. 

- If user explicitly specifies it and Globalnet is able to allocate those IPs, then this field will be same as Spec.GlobalIPs.

- If user explicitly specifies it but Globalnet **cannot** allocate those IPs, then this field will be empty and Status will be updated to Error (with the appropriate message).
- If user does not specify Spec.GlobalIPs, then this field includes the allocated globalIPs.

Reviewer: Can't we have the controller fill that field if it's omitted?

Author: We can do that, but if controller updates it, we may not know if the globalIP was requested by the user or auto-allocated by Globalnet later when we query the object. If you think that it does not really matter, I can remove it from Status.

@mangelajo your thoughts please.

Reviewer: Well, we can theoretically have an annotation that marks if it was auto allocated. I don't think it matters much though, but maybe I'm wrong?

Author: Lets hear from @mangelajo :-)

Reviewer 2: Same as in my other question, what happens if we want multiple ips?

Reviewer 2: The new GlobalnetIP + NumGlobalnetIps looks good to me.

Author: @mangelajo So shall we retain the allocated globalIPs (as a list) in the Status field (i.e., Status.GlobalIPs) or shall the controller just update Spec.GlobalIP and Spec.NumGlobalIPs itself (and remove Status.GlobalIPs)

Reviewer 3: We can use Spec.GlobalIP for user allocated and Status.GlobalIP for actual allocated. So Spec contains what was configured and Status contains what is actually allocated.

Author: In the current CRD Status.GlobalIPs is a list. So if four globalIPs are allocated (assuming that user specified Spec.numGlobalIPs=4), then Status.GlobalIPs will be as follows

```
   status:
     globalIPs:
      - 169.254.0.100
      - 169.254.0.101
      - 169.254.0.102
      - 169.254.0.103
```

Do you think we can rather use something like this (which reduces the size of the object) as the IPs are contiguous?
```
   status:
     globalIPs: "169.254.0.100-169.254.0.103"
```

Reviewer: > @mangelajo So shall we retain the allocated globalIPs (as a list) in the Status field (i.e., Status.GlobalIPs) or shall the controller just update Spec.GlobalIP and Spec.NumGlobalIPs itself (and remove Status.GlobalIPs)

This went unaddressed..

> Do you think we can rather use something like this (which reduces the size of the object) as the IPs are contiguous?
> 
> ```
>    status:
>      globalIPs: "169.254.0.100-169.254.0.103"
> ```

If we're keeping `status` and the `GlobalIP` and `NumGlobalIPs` fields are optional then I'd expect them to be mirrored in the `status` field, instead of going with a different format.

Author: > If we're keeping `status` and the `GlobalIP` and `NumGlobalIPs` fields are optional then I'd expect them to be mirrored in the `status` field, instead of going with a different format.

I'm fine with this proposal if it's okay for others.

Author: Let's re-visit this after we have an agreement on removing the `Spec.GlobalIP`. If we remove Spec.GlobalIP, then we can reflect the allocated globalIPs in `Status` field as `Status.GlobalIP` and `Status.NumGlobalIPs`

Reviewer 2: I would provide a list of IPs, otherwise it's harder to parse for the consumers, and if you have fragmentation you could end with things like:

169.254.0.1-169.254.0.3,169,169.254.0.9

or stuff like that.

I think a list is pretty straight forward, and not that consuming (let's avoid trying to optimize where is likely not necessary), I think usability is more important.

Reviewer 2: Also, if we are going to remove the ability to request specific IPs, you need to talk to @mkimuram it was his request.

I think it's important-ish, because such IPs will be used as part of firewall rules which could be not automated to track this API. Of course requesting an specific IP can still fail because such IP is not available. But.. if we try to always use the lower-ips when selecting automatically, the admins could use the end of the range for manual selection.

Author: I'm afraid, we cannot support Fragmentation with iptables. The IPAddresses should be contiguous.

Reviewer 2: One day (probably not that far) we will be doing only nftables. (we are already doing that in the background), or ovn rules...  may be we should not tailor the API to the implementation and make the API free of the details of the implementation as much as we can.

Reviewer 2: In OVN for example it's more restrictive (CIDRs instead of IP ranges..)

Reviewer: > One day (probably not that far) we will be doing only nftables. (we are already doing that in the background), or ovn rules... may be we should not tailor the API to the implementation and make the API free of the details of the implementation as much as we can.

But in that case the Spec needs to change no? Or is it enough just to ask for "X global ips" and the implementation will allocate what it can?

> In OVN for example it's more restrictive (CIDRs instead of IP ranges..)

That's interesting BUT is the globalnet processing tied to the CNI? I thought we're using iptables because it fits our needs, but not necessarily that it has to use the CNI.

Author: Currently, it is a List and if the backend implementation supports non-contiguous globalIPs, we can still show that to the user. So, its not tied to implementation. 

@mkolesnik, Globalnet implementation is mostly tied to kubeproxy iptables mode. Most of the CNIs we have today use kubeproxy iptables. OVN is an exception which programs its own flows to support Services.

    Spec               ClusterGlobalEgressIPSpec `json:"spec"`

    // Observed status of ClusterGlobalEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type ClusterGlobalEgressIPSpec struct {
    // The number of globalIP's requested. Globalnet Controller will allocate the requested
    // number of contiguous GlobalIPs for this ClusterGlobalEgressIP object.
    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
    // +optional
    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
}

type GlobalEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalEgressIP object.
    Spec               GlobalEgressIPSpec `json:"spec"`

    // Observed status of GlobalEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type GlobalEgressIPSpec struct {
    // The number of globalIP's requested. Globalnet Controller will allocate the requested
    // number of contiguous GlobalIPs for this GlobalEgressIP object.

>     // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.

Reviewer 5: If unspecified, it will be 0 in code unless we use *int. What if the user explicitly makes it 0? Also I assume we'll handle updates. If so what's the behavior when expanding and contracting?

Author: Good point Tom. IMHO 0 does not make sense as its as good as not creating the CR. We should be able to support updates. For 0.10 release, we can decide whether we want to support updates or defer to later release. From an implementation POV it should not be too complex to support updates.

    // +optional
    NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`

    // Selects the pods whose label match the definition. This is an optional field and
    // in case it is not set, it results in all the Pods selected from the namespace in which the
    // CR is created.
    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

const (
    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"

    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
    GlobalEgressIPError GlobalEgressIPStatus = "Error"
)

type EgressIPStatus struct {
    // Status is one of {"Success", "Error"}
    Status GlobalEgressIPStatus `json:"status,omitempty"`

    // +optional
    Message *string `json:"message,omitempty"`

    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
    GlobalIPs []string `json:"globalIPs"`
}
```

When Submariner Globalnet is deployed in a Cluster, it auto-allocates the configured number of
`globalIPs` and programs the necessary egress rules on the active Gateway node of the Cluster.

> This is done internally by creating `ClusterGlobalEgressIP` CR object with the name `cluster-default`.

Reviewer 2: If the admin creates any other ClusterGlobalEgressIP that we are going to ignore we must put it in error status and explain that they must use the 'cluster-default' or final name we  use.

to create `GlobalEgressIP` objects.

### Backward Compatibility

Inorder to support the newer implementation, a combination of ipSets with multiple IPTable chains
will be used. Also, we will internally use configMaps to maintain the state of allocated GlobalIPs
and get rid of annotations on the objects. Because of this, backward compatibility cannot be maintained
with the existing implementation of Globalnet.

### User Stories

1.User has two (or more) clusters with Overlapping CIDRs and requires the Pods in one
  Cluster be able to discover and talk to exported Services.

  To achieve this use-case, user can simply deploy Submariner Globalnet on all the participating
  clusters. By default, a globalIP (or a set of globalIPs) will be assigned at the Cluster level
  which will be used as egressIPs for inter-cluster communication and necessary Services can be
  exported.

2.In a Globalnet deployment user wants all the Pods in a namespace `ns1` to use a unique globalIP
  instead of the globalIP assigned at the cluster level.

  To achieve this, one can apply the following CRD in the namespace `ns1`.

> 

Reviewer 2: What about a GlobalIngressIP for reporting the assigned IPs to ExportedServices? (instead of annotations to ExportedServices?), I was thinking of the issues we had in the past with annotations being removed by the deployment tools sometimes.

Author: Yes Miguel, I'd prefer to use GlobalIngressIP CRD with minimal fields instead of annotating the Service/ServiceExport objects based on the past experience.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalEgressIP
   metadata:
     name: ns-egressip
     namespace: ns1
   spec:
     numGlobalIPs: 1
```

3.In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace
  to use a unique globalIPs as egressIP for cross-cluster communication.

  To achieve this, one can apply the following CRD in the namespace `ns1`.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalEgressIP
   metadata:
     name: db-pods
     namespace: ns1
   spec:
     podSelector:
         matchLabels:
           role: db
     NumGlobalIPs: 2
```

5.In a Globalnet deployment, user wants a specific service to be made available to remote clusters.
  For this, user can create a ClusterIP Service and export the Service. Globalnet will allocate a

>   globalIP for ingress connectivity such that any external applications can connect to this Service.

Reviewer 5: This seems a duplicate of 1.

Author: 1 talks about EgressIP support whereas this is about IngressIP support.

will be used. Also, we will internally use configMaps to maintain the state of allocated GlobalIPs
and get rid of annotations on the objects. Because of this, backward compatibility cannot be maintained
with the existing implementation of Globalnet.

### User Stories

1. User has two (or more) clusters with Overlapping CIDRs and requires the Pods in one
   Cluster be able to discover and talk to exported Services.

   To achieve this use-case, user can simply deploy Submariner Globalnet on all the participating
   clusters. By default, a globalIP (or a set of globalIPs) will be assigned at the Cluster level
   which will be used as egressIPs for inter-cluster communication and necessary Services can be
   exported.

2. In a Globalnet deployment user wants all the Pods in a namespace `ns1` to use a unique globalIP
   instead of the globalIP assigned at the cluster level.

   To achieve this, one can apply the following CRD in the namespace `ns1`.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalEgressIP
   metadata:
     name: ns-egressip
     namespace: ns1
   spec:
     numGlobalIPs: 1
```

> 3. In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace

Author: @mkolesnik, if I give an additional space between "3. In", there is a linter issue

```
submariner/globalnet-enhancement2-0.md: 213: MD029/ol-prefix Ordered list item prefix [Expected: 1; Actual: 3; Style: 1/1/1]
```

Reviewer: It seems to be because you have a blank line above

## Globalnet 2.0: Enhancements to Globalnet Implementation

## Summary

> Currently, Submariner Globalnet implementation in [release 0.8](https://github.com/submariner-io/releases/releases/tag/v0.8.0)

Reviewer: ```suggestion
Currently, Submariner Globalnet implementation in [release 0.9](https://github.com/submariner-io/releases/releases/tag/v0.9.0)
```

nit :)

annotates every Pod, Service and Node in the cluster with a globalIP to support Pod to remote
Service connectivity in the following manner.

Once a Pod is scheduled and enters Running state, it is annotated with a globalIP. To support
the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
the cluster where the source IP of the Pod is modified to the corresponding globalIP. This
translation enables us to uniquely identify the traffic from the Pod, when it reaches the
destination cluster.

Similarly, a ClusterIP Service is annotated with a globalIP so that remote Pods can communicate
with it using its globalIP. Globalnet programs the necessary ingress rules on the active Gateway
node of the cluster, where the service resides. The ingress rules enable any incoming traffic
destined to the globalIP of the Service to be redirected to the corresponding ClusterIP Service.

In order to support connectivity from the Host Network to the remote services, Globalnet also

> annotates every node in the cluster with a unique globalIP.

Reviewer: BTW just noticed this mentioned here but not anywhere else in the EP.

Is this behavior maintained?

Author: Yes, its maintained.

Reviewer: Probably should be explicitly mentioned since otherwise it seems the whole of Globalnet v1 is being replaced and this functionality will be lost.

While this works fine for small clusters, where the number of Pods and Services are limited, for
large clusters it poses scalability and performance issues as we would eventually run out of
globalIPs because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod

> connectivity as well as Headless Services are not supported.

Reviewer: ```suggestion
globalIPs because of annotating every Pod/Service/Node object. Also, in this design, 
Headless Services are not supported.
```

AFAICT pod to pod is not going to be supported by this proposal either, so I wouldn't mention it at all.

Author: Its supported via Headless Services

## Proposal

This enhancement proposes the following changes to make Globalnet implementation scale better.

Egress traffic:

* By default, every cluster will be assigned a configurable number of globalIPs which will be
  used as egress IPs for any cross-cluster communication. Support for more than a single globalIP
  at the cluster level is done to avoid ephemeral port exhaustion issues.
* Provision to allocate a set of globalIPs at the namespace/project level which takes precedence
  over the globalIPs allocated to the cluster. All Pods in the namespace will use the same globalIPs
  as EgressIPs.

> * Provision to allocate globalIPs to a set of Pods in a selected namespace or cluster-wide.

Reviewer: ```suggestion
* Provision to allocate globalIPs to a set of Pods in a namespace.
```

  Such Pods will use globalIPs that take precedence over any globalIPs allocated at the default
  namespace level.

Ingress traffic:

* Instead of assigning a globalIP every ClusterIP Service in the cluster, only exported Services

>   will be assigned a globalIP. Globalnet internally creates a GlobalIngressIP CR for every

Reviewer: Can you add details of the proposed CRD `GlobalIngressIP`?

Author: Is it Okay if we discuss the CRD as part of implementation PR that adds support for Services? JFYI, the `GlobalIngressIP` is internally created by Globalnet when Service is exported and is not expected to be created by the users.

Reviewer: > Is it Okay if we discuss the CRD as part of implementation PR that adds support for Services?

I think defining the CRD (even if internal) should help understand the ingress side of the proposal which is a bit vague IMHO, eg is it one IP or more? How is it connected to service? How is it designed for headless services?

> JFYI, the `GlobalIngressIP` is internally created by Globalnet when Service is exported and is not expected to be created by the users.

What would happen if a user creates one though?
Or would that be blocked by RBAC?

Author: > I think defining the CRD (even if internal) should help understand the ingress side of the proposal which is a bit vague IMHO, eg is it one IP or more? How is it connected to service? How is it designed for headless services?

I'd imagine that for every exported ClusterIP Service and Headless Service, we will internally create a GlobalnetIngress object in the same namespace with the same name. 

Regarding the number of IPs:
For an exported ClusterIP Service, there will only be a single globalIP as ingress IP.
For a Headless Service, for every backend pod, we will have a unique globalIP as ingress IP.

@vthapar please let me know if this works or if you think we will have to model it differently.

> What would happen if a user creates one though?
> Or would that be blocked by RBAC?

Ideally, I will expect it to be blocked by RBAC, but we will have more clarity when we start working on the implementation.

Reviewer: Hence why I think it's best to define the CRD here just you're defining the other CRDs.

Another way of looking at it - why describe the other CRDs but not this one?

It's definitely an API and more a design detail than implementation detail.

Reviewer 3: @mkolesnik @sridhargaddam I am leaning towards a GlobalIngress object for each pod in case of headless service instead of tying it to exported headless service. Way I see it, it is similar to user allocating/requesting global IP to a pod, just that in case of headless service we are doing it when a service is allocated. We can add labels to mark which services they belong to optimize lookups.

Reviewer 4: If we define `GlobalIngress` object in a way that align to k8s `Service` and `Endpoints` object, it would be like:

<details>

```
type GlobalNetService struct {
	metav1.TypeMeta
	// +optional
	metav1.ObjectMeta

	// Spec defines the behavior of a service.
	// +optional
	Spec GlobalNetServiceSpec

	// Status represents the current status of a service.
	// +optional
	Status GlobalNetServiceStatus
}

type GlobalNetServiceSpec struct {
	// GlobalIP is the IP address of the service from outside a cluster
	// +optional
	GlobalIP string
}

type GlobalNetServiceStatus struct {
	// Current GlobalNetService condition
	// +optional
	Conditions []metav1.Condition
}

type GlobalNetEndpoints struct {
	metav1.TypeMeta
	// +optional
	metav1.ObjectMeta

	// The set of all endpoints is the union of all subsets.
	Subsets []GlobalNetEndpointSubset
}

type GlobalNetEndpointSubset struct {
	// GlobalIP is the IP address of the service from outside a cluster
	GlobalIP string

	// EndpointAddress is a tuple that describes single IP address.
	EndpointAddress v1.EndpointAddress

	Ports           []v1.EndpointPort
}
```

</details>

If these resources are created by controller when a `ServiceExport` is created by user and are reconciled on updates in `Service`, `Endpoints`, and `Pod`, it will cover ingress of exported `Service` for both normal and headless case. In addition, it will include definition of egress for headless service.

I'm also thinking about `GlobalEgress` can also be aligned to this (`GlobalEgress` will be just sets of `GlobalNetEndpointSubset` that has empty `Ports`, despite that there's no such concept in k8s, like `Service` without Ports).

Reviewer 3: @mkimuram ServiceExport/Import are somewhat analogous to `GlobalNetService*` objects you're proposing. `GlobalnetIngress` object I was thinking of similar to what you've proposed as `GlobalnetEndpoints` but wanted to go with single object per pod than a list of endpoints. Pros of a separate object per pod is that any change in pods will result in a add/delete operation instead of udates on a common object. Since we may need to sync these objects across clusters, it will avoid conflicts. Besides, even if we go with your proposal, we will have to mirror `EndpointSlice` than `Endpoint` for scalability reasons.

Reviewer: @vthapar yes that sounds reasonable (GlobalIngress object for each pod in case of headless service)

Since this will be managed by the globalnet controller automatically, it shouldn't be hard for it to ensure exported headless service pods have the CR.

This is why I asked that we define this CRD in the EP as well, so that we can understand the proper design (as it is an API in the end).

Author: Many thanks @mkimuram for the detailed CRD proposal. I had a meeting with @vthapar and we discussed the implications of using it with Lighthouse and felt that creating individual GlobalIngressIPs objects (i.e., a single object for every exported ClusterIP service and one for each of the backend Pod for an Headless Service) will be easier to maintain. I hope you are okay with the proposal? I'm updating this PR to show how the CRD looks like.

Reviewer 4: > I hope you are okay with the proposal? 

Yes, it makes sense.

  exported service.

> * For Headless Services, necessary ingress rules will be programmed to support connectivity to

Reviewer: BTW how is this modelled for headless service?

Are they expected to use same `GlobalIngressIP` CRD or something else?

Author: I'd like to discuss with @vthapar on how to model the `GlobalIngressIP` CRD so that we can use the same CRD for both ClusterIP services and Headless Services. Hence suggesting that we can probably review that as part of the respective PR.

  assigned via GlobalEgressIP takes precedence over individual Pod globalIPs as EgressIP.

Both EgressIP as well as IngressIPs will be allocated from the same globalCIDR Pool used in
the Cluster. Globalnet will not support an EgressIP/IngressIP that is outside of globalCIDR
Pool.

![GlobalnetTopology](./images/globalnet-topology.png)

If the external application in a remote Cluster is interested in consuming the exported
Service, it will have to reach the globalIP assigned to the exportedService and the response
from the backend Pods will have the source as the globalIP of the exportedService.

On the other hand, when the backend Pod tries to talk to the external application in remote
Cluster, the external application will see that traffic is coming with a globalIP that is
assigned either at the Cluster level, or namespace level, or the one requested by the user
at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
for both ingress and egress traffic can create Headless Services in front of them and should

> avoid requesting explicit globalIP that match Headless Service backend Pods.

Reviewer: This bit is problematic since according to the diagram the "namespace IP" takes precedence over the ingress IP, but what if as a user I want the specific pods for a service to use the same address? Then I can't even have namespace level assignment that catches all other pods (or I have to use some excluding selector or something?)

As a user I think it would be simpler is the order or precedence for egress was:
1. `GlobalEgressIP` with pod selector that matches the pod's IP
2. `GlobalIngressIP` (or whatever for headless service)
3. `GlobalEgressIP` with empty selector
4. `ClusterGlobalEgressIP`

Then you still have the possibility to provide a-synch behavior while the default is sync. This is also, AFAICT, what is explained in user story 6.

**OTOH for simplicity's sake, and if we want to mimic K8s as close as possible, we should do the following:**
* Headless service IPs always take precedence (matches [1] AFAICT)
* Cluster service IPs are only used for ingress

[1] https://kubernetes.io/docs/concepts/services-networking/service/#headless-services
> For headless Services, a cluster IP is not allocated, kube-proxy does not handle these Services, and there is no load balancing or proxying done by the platform for them. How DNS is automatically configured depends on whether the Service has selectors defined:

Author: > This bit is problematic since according to the diagram the "namespace IP" takes precedence over the ingress IP, but what if as a user I want the specific pods for a service to use the same address? Then I can't even have namespace level assignment that catches all other pods (or I have to use some excluding selector or something?)
> 
> As a user I think it would be simpler is the order or precedence for egress was:
> 
> 1. `GlobalEgressIP` with pod selector that matches the pod's IP
> 2. `GlobalIngressIP` (or whatever for headless service)
> 3. `GlobalEgressIP` with empty selector
> 4. `ClusterGlobalEgressIP`

Correct @mkolesnik. As described textually, this was the idea we had in mind, but I forgot to update the diagram. Thanks.
 
> **OTOH for simplicity's sake, and if we want to mimic K8s as close as possible, we should do the following:**

True, this will be simpler to support. But this was a requirement from external network use-cases @mkimuram @mangelajo 
 
> * Headless service IPs always take precedence (matches [1] AFAICT)
> * Cluster service IPs are only used for ingress
> 
> [1] https://kubernetes.io/docs/concepts/services-networking/service/#headless-services
> 
> > For headless Services, a cluster IP is not allocated, kube-proxy does not handle these Services, and there is no load balancing or proxying done by the platform for them. How DNS is automatically configured depends on whether the Service has selectors defined:

Reviewer: > True, this will be simpler to support. But this was a requirement from external network use-cases @mkimuram @mangelajo

Is the requirement documented here (in the EP or a discussion)? If so I'd like to have a read to understand why it's needed.

To me this simple approach would be best since it follows K8s and also is very simple to understand and work with, so understanding the external networking requirement is important.

Author: @mkolesnik yes it was discussed already in this EP. Please see - https://github.com/submariner-io/enhancements/pull/22#discussion_r571030250

Reviewer 4: > But this was a requirement from external network use-cases

I think that such a requirement of controlling source IPs exists even without external network use case. For example, with cross-cluster use case, users will require it when they would like to use network policies. Inside a single cluster, they can use namespace selector or pod selector, howerver outside the cluster, they need to rely on ingress IPs.

Reviewer: > > But this was a requirement from external network use-cases
> 
> Inside a single cluster, they can use namespace selector or pod selector, howerver outside the cluster, they need to rely on ingress IPs.

We have a separate project to tackle cross cluster network policy and it'll support such use cases in an abstract way, so we shouldn't base our design on that.

Anyhow, if the effective order is going to be that `GlobalIngressIP` takes precedence over `GlobalEgressIP` with empty selector I think we can live with that, even though I don't agree that it's the best solution to the problem that has been presented.

Author: Thanks @mkolesnik and @mkimuram

    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
    GlobalEgressIPError GlobalEgressIPStatus = "Error"
)

type EgressIPStatus struct {
    // Status is one of {"Success", "Error"}
    Status GlobalEgressIPStatus `json:"status,omitempty"`

    // +optional
    Message *string `json:"message,omitempty"`

    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
    GlobalIPs []string `json:"globalIPs"`
}
```

When Submariner Globalnet is deployed in a Cluster, it auto-allocates the configured number of
`globalIPs` and programs the necessary egress rules on the active Gateway node of the Cluster.
This is done internally by creating `ClusterGlobalEgressIP` CR object with the name `cluster-default`.
Users can query the CR to check the allocated GlobalIPs at the Cluster level. If admin creates
additional ClusterGlobalEgressIP CRs that match the Cluster scope, the status field of those CRs will
be marked as error with appropriate message. Basically only `cluster-default` object will be supported.

Only an admin user can create the `ClusterGlobalEgressIP` object and regular users will only be able

> to create `GlobalEgressIP` objects.

Reviewer: ```suggestion
to create `GlobalEgressIP` and `GlobalIngressIP` objects.
```

Author: I'm afraid No. `GlobalIngressIP` will be created/used internally within the Globalnet implementation when a user exports a Service. We don't expect users to create it manually.

Reviewer: I've read the section as "permissions (RBAC)" not as "when/how is it created".

Also even though users aren't expected to create it, they'll still be able to query it no?

Author: Yes, user (like Lighthouse) will be able to query but will not be able to create it. Updated the section in Line46

### Backward Compatibility

> Inorder to support the newer implementation, a combination of ipSets with multiple IPTable chains

Reviewer: ```suggestion
In order to support the newer implementation, a combination of ipSets with multiple IPTable chains
```

nit :)

will be used. Also, we will internally use configMaps to maintain the state of allocated GlobalIPs
and get rid of annotations on the objects. Because of this, backward compatibility cannot be maintained
with the existing implementation of Globalnet.

### User Stories

> 1.User has two (or more) clusters with Overlapping CIDRs and requires the Pods in one

Reviewer: ```suggestion
1. User has two (or more) clusters with Overlapping CIDRs and requires the Pods in one
```

Reviewer: This hasn't been fixed.

It renders pretty poorly in the rich view, either use an ordered list properly or use sub headings for the different user stories.

Compare how this renders:
![Screenshot from 2021-05-10 13-31-46](https://user-images.githubusercontent.com/51078027/117646399-60aa9400-b194-11eb-974f-1909cd4178ff.png)

With how work items section renders (properly):
![Screenshot from 2021-05-10 13-31-57](https://user-images.githubusercontent.com/51078027/117646431-6acc9280-b194-11eb-80ad-86a19f8ee058.png)

  Cluster be able to discover and talk to exported Services.

  To achieve this use-case, user can simply deploy Submariner Globalnet on all the participating
  clusters. By default, a globalIP (or a set of globalIPs) will be assigned at the Cluster level
  which will be used as egressIPs for inter-cluster communication and necessary Services can be
  exported.

> 2.In a Globalnet deployment user wants all the Pods in a namespace `ns1` to use a unique globalIP

Reviewer: ```suggestion
2. In a Globalnet deployment user wants all the Pods in a namespace `ns1` to use a unique globalIP
```

  instead of the globalIP assigned at the cluster level.

  To achieve this, one can apply the following CRD in the namespace `ns1`.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalEgressIP
   metadata:
     name: ns-egressip
     namespace: ns1
   spec:
     numGlobalIPs: 1
```

> 3.In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace

Reviewer: ```suggestion
3. In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace
```

  to use a unique globalIPs as egressIP for cross-cluster communication.

  To achieve this, one can apply the following CRD in the namespace `ns1`.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalEgressIP
   metadata:
     name: db-pods
     namespace: ns1
   spec:
     podSelector:
         matchLabels:
           role: db
     NumGlobalIPs: 2
```

> 5.In a Globalnet deployment, user wants a specific service to be made available to remote clusters.

Reviewer: ```suggestion
4. In a Globalnet deployment, user wants a specific service to be made available to remote clusters.
```

  For this, user can create a ClusterIP Service and export the Service. Globalnet will allocate a
  globalIP for ingress connectivity such that any external applications can connect to this Service.

```shell
  kubectl -n default create deployment nginx --image=nginxinc/nginx-unprivileged:stable-alpine
  kubectl -n default expose deployment nginx --port=8080
  subctl export service --namespace default nginx
```

> 6.In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace

Reviewer: ```suggestion
5. In a Globalnet deployment, user wants a selected Pod (or set of Pods) in a particular namespace
```

  to use a unique globalIP both as ingress and egressIP for cross-cluster communication.

  To achieve this, user can create a Headless Service (backed by a Deployment or StatefulSets) and
  Globalnet will ensure that all the backend Pods that belong to the Headless Service are assigned
  a globalIP. User is also expected **not** to create any GlobalEgressIP object which selects
  the backendPods that match the Headless Service.

![HeadlessService](./images/globalnet-headless-svc.png)

  In the above example, each of the http pods will get its own globalIP which can be used for
  both ingress and egress communication.

>   However, if the user wants egress traffic from the http Pods to carry a unique globalIP,

Reviewer: This conflicts with the order stated in the diagram.

Author: No, it doesn't. Are you referring to `/images/globalnet-svc.png` at L271 image or something else?

  its possible by explicitly requesting a GlobalEgressIP with PodSelectors matching
  Headless Service Pods as shown below.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalEgressIP
   metadata:
     name: http-pods
     namespace: myproject
   spec:
     podSelector:
       matchLabels:
         app: http
```

![Headless Service with EgressIP](./images/globalnet-svc.png)

## Alternatives

> It is possible to support Pod to Pod communication even with Globalnet. However, to support this, it

Reviewer: This is only place where it mentions pod to pod, but it seems that Globalnet v1 supports pod to pod better and the new proposal won't have pod to pod communication unless you're using headless service, and are very careful with your `GlobalEgressIP` allocations.

Author: In general we dont need Pod to Pod reachability. As you know, the standard way to connect is for a Pod to discover a Service and connect to the Service. As there was no strong use-case we did not support Pod to Pod with Globalnet V1. 

In V2, we will be adding support via Headless Services. If you think this sentence needs to be rephrased, please let me know.

Reviewer: > In general we dont need Pod to Pod reachability. As you know, the standard way to connect is for a Pod to discover a Service and connect to the Service. As there was no strong use-case we did not support Pod to Pod with Globalnet V1.

We didn't support it explicitly, but it was easily doable since every pod gets a global IP so in practice it's supported.

> In V2, we will be adding support via Headless Services. If you think this sentence needs to be rephrased, please let me know.

Hence I think calling it "pod to pod" is confusing since it implies that you can reach pods directly.
Since only "pod to pod" expected is via headless services, let's refer only to this terminology.

Author: Okay, let me rephrase the sentence.

the cluster where the source IP of the Pod is modified to the corresponding globalIP. This
translation enables us to uniquely identify the traffic from the Pod, when it reaches the
destination cluster.

Similarly, a ClusterIP Service is annotated with a globalIP so that remote Pods can communicate
with it using its globalIP. Globalnet programs the necessary ingress rules on the active Gateway
node of the cluster, where the service resides. The ingress rules enable any incoming traffic
destined to the globalIP of the Service to be redirected to the corresponding ClusterIP Service.

In order to support connectivity from the Host Network to the remote services, Globalnet also
annotates every node in the cluster with a unique globalIP.

While this works fine for small clusters, where the number of Pods and Services are limited, for
large clusters it poses scalability and performance issues as we would eventually run out of
globalIPs because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
connectivity as well as Headless Services are not supported.

## Proposal

This enhancement proposes the following changes to make Globalnet implementation scale better.

> Egress traffic:

Reviewer: So essentially, egress IPs aren't necessarily related to the `GlobalCidr` field?

Author: I'm afraid no. EgressIPs will be picked from the GlobalCIDR allocated to the cluster. 

It is clarified in Line 34 and L113. If you think its not clear, we can update the description.

Reviewer: Yeah it's not very clear that both ingress and egress IPs will be taken from the same pool. In fact, I was under the impression the GlobalCIDR will only be used for the ingress IPs.

Moreover, I think in this proposal it would still be possible to allocate an IP outside this CIDR (though probably this shouldn't be allowed as otherwise we'll need to support routing them as well).

Author: Okay, I'll add some notes to clarify that point.

We do not support allocation of globalIP that is outside of globalCIDR Pool used in the cluster as it has implications on routing as well.

* By default, every cluster will be assigned a configurable number of globalIPs which will be
  used as egress IPs for any cross-cluster communication. Support for more than a single globalIP
  at the cluster level is done to avoid ephemeral port exhaustion issues.
* Provision to allocate a set of globalIPs at the namespace/project level which takes precedence
  over the globalIPs allocated to the cluster. All Pods in the namespace will use the same globalIPs
  as EgressIPs.
* Provision to allocate globalIPs to a Pod (or a set of Pods) from the selected namespaces or
  from any namespace. Such Pods will use globalIPs that take precedence over any globalIPs allocated
  at the default namespace level.

> Ingress traffic:

Reviewer: As per my understanding, these IPs need to be distinct from the egress IPs, right?

Author: Correct. Apart from Services, only Pods backed by Headless Services get an IngressIP and the behavior is explained in use-case 5.

Reviewer: What I meant is that these IPs here shouldn't overlap with the IPs that are assigned for egress.

E.G lets say an egress ip is assigned to pods by a selector, this IP shouldn't be used by any ingress IP allocation or it'll lead to confusion.

I think generally that we need provisioning to make sure that global IPs aren't colliding in a cluster (and across clusters).

Author: > What I meant is that these IPs here shouldn't overlap with the IPs that are assigned for egress.

I'm afraid, no. The EgressIPs and IngressIPs are from the same globalnetCIDR Pool allocated to the Cluster. Even in the current Globalnet Implementation, we annotate Pods (aka EgressIP) and Services (aka IngressIP) from the same Pool.

> I think generally that we need provisioning to make sure that global IPs aren't colliding in a cluster (and across clusters).

Same globalIP will **not** be allocated to a Pod/Namespace/Cluster (aka Egress) and a Service (aka Ingress).
So they will not collide in the local cluster (as well as across clusters because other clusters use a totally different globalnetCIDR Pool).

For Headless Services, which does not have ClusterIP and LH/DNS advertises IPs of backendPods, the globalIPs assigned to such backendPods will be used for both Ingress/Egress as explained in use-case 5 (please look at the corresponding diagram, it will help in visualizing the scenario). 

This is done as there could be some specific use-cases where External clients want to have the same IP for both ingress/egress (which is possible with Vanilla K8s when using Headless Services). So we are supporting the use-case even in the new Globalnet design to be aligned with the behavior of Vanilla K8s. For non Headless Services, the EgressIPs and IngressIP will be different though.

Reviewer: > > I think generally that we need provisioning to make sure that global IPs aren't colliding in a cluster (and across clusters).
> 
> Same globalIP will **not** be allocated to a Pod/Namespace/Cluster (aka Egress) and a Service (aka Ingress).
> So they will not collide in the local cluster (as well as across clusters because other clusters use a totally different globalnetCIDR Pool).
> 

Sure, what I mean is since users can request specific IPs now we'll need to make sure not only the automatic allocation of IPs doesn't collide, but  also the user requested IPs don't collide (and won't be automatically allocated).

E.G if an IP (or range) is requested and already allocated (or partially allocated for a range) then the resource should be in error state and the IP(s) shouldn't be allocated to it.

Author: Yes, it will be taken care of while allocating globalIPs. We are on the same page.

    Spec               EgressIPSpec `json:"spec"`

    // Observed status of GlobalnetEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type EgressIPSpec struct {
    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
    // K8s objects, then it will be allocated and the status field will be updated accordingly.
    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
    // If this field is not specified, a single globalIP will be allocated.
    // This field cannot be changed through updates.
    // +optional
    GlobalIPs  []string `json:"globalIPs,omitempty"`

    // Selects the Namespaces to which this GlobalnetEgressIP object applies.
    // If an empty namespaceSelector: {} is configured, it selects all the Pods in all namespaces.
    // On the other hand, if you omit specifying namespaceSelector, it does not select any
    // namespaces and selects only Pods from the namespace where the GlobalnetEgressIP is
    // deployed to.
    // +optional

>     NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`

Reviewer: So in fact a CR with empty namespace and pod selectors is a cluster wide policy then?

Author: Good point. As commented in this PR, by default when Globalnet is deployed in a Cluster, it allocates a single (or configurable no of) globalIPs and there was no provision to request a specific globalIP at the cluster level. 

@mangelajo @mkolesnik I did not think about this configuration. Do you think we should allow this configuration and apply the Requested GlobalIPs (which is an optional field btw) cluster wide? or shall we simply return error?

Reviewer: I think it'll be consistent if we allow to define a cluster level policy by creating one with no namespace/pod selectors, you'll have to have this information stored somewhere anyhow and this way its logically consistent.

Author: Okay. My idea behind auto-allocation of **globalIPs at Cluster level** is to have the basic connectivity working as soon as Submariner is deployed instead of expecting the user to create any GlobalnetEgressIP objects.

Can we do something like this? As soon as Globalnet is deployed in a cluster, it internally creates a GlobalnetEgressIP object with no namespace/pod selectors (i.e., applies to cluster) with a well defined name. Users can query the object and see the globalIP assigned at the cluster level. But users cannot modify/delete it in the initial implementation.

If user starts to create new objects (that apply at cluster level) there will be no guarantee which of the globalIP address will be used.

Reviewer: I'd create a default one with some reasonable default number of IPs (10?) when submariner is installed. We can also expose the pool size as an option for the installation, so that the users can specify it if they want.

I'd allow users to create a new one and delete the old one which is created by default. I don't see a specific problem with this. It's the same as allowing any other creation or deletion of resources.

If multiple "cluster" policies exist we shouldn't commit to which one takes precedence, in any case we could program multiple iptable rules and the first one will catch the packets, so it's really not a problem to have even multiple policies as long as users are informed that only one will be in effect at any given time.

Reviewer 2: I like the idea of having the default one created. Then the cluster globalIPs are more explicit and can be consulted.

Like the "default" namespace.

improve the overall performance of the solution.

### Cons

A limitation with the proposed approach is that we will lose the unique identity for **every**
Pod which is supported today. For users that intend to use Network Policies across Clusters or
for debugging purposes, this can be mitigated by creating the necessary NamespaceGlobalEgressIP CRDs
which create unique globalIPs at the Namespace/Pod level.

### Caveats

In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
and the corresponding egress rules are programmed on the node, it was not able to communicate with
remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
will be closer to the Vanilla Submariner experience.

## Design Details

The following CRD is proposed to support the new design.

```Go
type ClusterGlobalEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of ClusterGlobalEgressIP object.

>     Spec               ClusterEgressIPSpec `json:"spec"`

Reviewer: Shouldn't this be `ClusterGlobalEgressIPSpec`?

Alternatively we can rename the `ClusterGlobalEgressIP` type to `ClusterEgressIP` which makes sense I guess since it is part of submariner globalnet?

Author: Ack, will change it to ClusterGlobalEgressIPSpec

    // Observed status of NamespaceGlobalEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type ClusterEgressIPSpec struct {
    // The number of globalIP's requested. Globalnet Controller will allocate the requested
    // number of contiguous GlobalIPs for this ClusterGlobalEgressIP object.
    // If unspecified, NumGlobalIPs defaults to 1.
    // +optional

>     NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`

Reviewer 2: I still think that the ability to request specific IPs optionally is good (RequestedGlobalIPs or something like that, as a list), of course, that could fail if the IPs are not available.

@mkimuram ^ could you still work without that feature?

Reviewer: We didn't say it's not possible, but for now there's enough work to be done without this ability.

In the future we can always add it, as the proposed API doesn't forbid adding this field, should it be necessary.

Author: @mkimuram clarified in today's call that currently, it is fine to remove GlobalIP from the Spec and add it in a future release.

    // Selects the pods whose label match the definition. This is an optional field and
    // in case it is not set, it results in all the Pods selected from all Namespaces.
    // +optional

>     PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`

Reviewer: I'm not sure if this is a real use case on the cluster level.

Also it might lead to confusing situations when you update the default cluster policy to match only specific pods, what will the rest of the pods get if they match no policy?

Maybe I'm wrong and there is a use case for this on the cluster level, but AFAICT the namespace level allocations should cover use cases where specific pods need to be selected.

Author: I forgot to remove it while updating the commit. Will update in the next commit.

}

> type NamespaceGlobalEgressIP struct {

Reviewer 2: ```suggestion
type GlobalEgressIP struct {
```

I would suggest that we follow the convention of other models in K8s, Role vs ClusterRole, RoleBinding vs ClusterRoleBinding, etc.

In our case it would be GlobalEgressIP vs ClusterGlobalEgressIP

Also makes the name shorter.

Author: Ack. Will do, thanks for reviewing @mangelajo

    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of NamespaceGlobalEgressIP object.

>     Spec               NamespaceEgressIPSpec `json:"spec"`

Reviewer: Same here about the spec type name not matching the type name

    // Observed status of NamespaceGlobalEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type NamespaceEgressIPSpec struct {
    // The number of globalIP's requested. Globalnet Controller will allocate the requested
    // number of contiguous GlobalIPs for this NamespaceGlobalEgressIP object.
    // If unspecified, NumGlobalIPs defaults to 1.
    // +optional

>     NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`

Reviewer 2: Same question here about requesting specific ips.

Author: Based on the above point, I'm marking it as resolved. We will look into this in future.

    // Selects the pods whose label match the definition. This is an optional field and
    // in case it is not set, it results in all the Pods selected from the namespace in which the
    // CR is created.
    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

const (
    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"

    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
    GlobalEgressIPError GlobalEgressIPStatus = "Error"
)

type EgressIPStatus struct {
    // Status is one of {"Success", "Error"}

>     Status GlobalEgressIPStatus `json:"status,omitempty"`

Reviewer: This will show up as
```
status:
  status: Success/Error
```

A bit weird :)

Perhaps a better name is `Working` or something of the sort?

Reviewer 5: K8s is standardizing on `metav1.Condition` (introduced in 1.19 which we're bumping to) for such things. To follow other upstream resources, this would become `Conditions []metav1.Condition`. You define your own type strings: we can use "_Allocated_" and perhaps "_Conflict_" and whatever else.

Author: Okay as discussed in today's call, we can modify this section to reflect the actual implementation later. This Enhancement PR is mainly to give a high-level picture and not go too much into details associated with CRDs. @tpantelis please feel free to use metav1.Condition in the Status

    // +optional
    Message *string `json:"message,omitempty"`

>     // When more than a single GlobalIP is allocated, the starting GlobalIP allocated to this object.

Reviewer 2: I think a list of IPs is fine it's not an object that will be subject to changes, so the storage & API related traffic should not be significant, even if we had thousands of IPs. For example 2048 IPv4 would result in +32kB of data.

I would think differently if this type of object could be changing a lot, and subject to informers, etc.

Also, having the flexibility to provide non-contiguous IP would avoid fragmentation issues.

Reviewer: As stated, IPs must be contiguous blocks so this design fits the limitations.

What is the added benefit of storing a list?

Author: I think we also have to restrict the number of GlobalIPs that can be requested. So added a max limit of 10.
In most of the cases, it will could be a single GlobalIP or maybe few GlobalIPs, so a list might make it easy to parse for the users. @mkolesnik I hope you are okay to have it as a list.

Regarding contiguous or non-contiguous, since the allocation will now be done by Globalnet (and not user-requested), lets leave it to implementation. IPTables based Globalnet requires them to be Contiguous. OVN based Globalnet might be able to support non-contiguous (not sure).

    StartIP string `json:"globalIPs"`

    // When more than a single GlobalIP is allocated, the last GlobalIP (inclusive) allocated to this object.
    EndIP string `json:"globalIPs"`

    // The Namespaces to which the GlobalIPs are applied.
    Namespaces []string `json:"namespaces"`

    // The Pods to which the GlobalIPs are applied.

>     Pods []string `json:"pods"`

Reviewer 2: I think we don't need to provide this information, it could be helpful for debugging but it would create a lot of work on the API side, and kind of overrides my comments in  L170, I hadn't noticed this.

We can add debug traces to this, ... but think of services, they provide no list of backend pods , just the list of IPs via a separate object (Endpoint).

I propose we remove Namespaces & Pods to avoid too much API trashing which would be a scalability issue (lots of pods come and go)

Reviewer: I tend to agree about the pods part, namespaces probably dont churn as much.

IMHO it's better to annotate the pods (and maybe namespaces?) with the global policies in effect.

For first stage I'd remove these fields altogether though as I believe there's enough work as is without these aspects.
As Miguel proposed, this can be logged on the globalnet pod for now.

Author: Done, removed both of them. Infact I also added a note in L184 that we will no longer be using annotations on the Pods/Services as we are seeing issues with some services managed by certain controllers which periodically remove the annotation.

>     // The Namespaces to which the GlobalIPs are applied.
>     Namespaces []string `json:"namespaces"`
> 
>     // The Pods to which the GlobalIPs are applied.
>     Pods []string `json:"pods"`

Reviewer 2: ```suggestion
```

Reviewer 2: If we need this because we don't want to label/annotate pods, it could be a separate CRD just for reporting.  But I wouldn't keep it in a single monolithic CR

Author: Yes, we will not be annotating the Pods/Services and will use ConfigMap to store the allocation info.

In order to support connectivity from the Host Network to the remote services, Globalnet also
annotates every node in the cluster with a unique globalIP.

While this works fine for small clusters, where the number of Pods and Services are limited, for
large clusters it poses scalability and performance issues as we would eventually run out of
globalIPs because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
connectivity as well as Headless Services are not supported.

## Proposal

This enhancement proposes the following changes to make Globalnet implementation scale better.

Egress traffic:

* By default, every cluster will be assigned a configurable number of globalIPs which will be
  used as egress IPs for any cross-cluster communication. Support for more than a single globalIP
  at the cluster level is done to avoid ephemeral port exhaustion issues.
* Provision to allocate a set of globalIPs at the namespace/project level which takes precedence
  over the globalIPs allocated to the cluster. All Pods in the namespace will use the same globalIPs
  as EgressIPs.

> * Provision to allocate globalIPs to a Pod (or a set of Pods) from the selected namespaces or

Reviewer: ```suggestion
* Provision to allocate globalIPs to a a set of Pods from the selected namespaces or
```

Since the proposal is to use selectors, it's unlikely only one pod gets the IP. I find this needs to be clearer.

  from any namespace. Such Pods will use globalIPs that take precedence over any globalIPs allocated
  at the default namespace level.

Ingress traffic:

* Instead of annotating every ClusterIP Service in the cluster, only exported Services will be
  annotated with a globalIP.
* For Headless Services, necessary ingress rules will be programmed to support connectivity to
  the backend Pods. In a Headless Service use-case, every backend Pod will be annotated with
  a unique globalIP as an Ingress IP. Clients can discover the associated globalIP using
  Lighthouse and can connect to the globalIP assigned to the backend pods.
  For Pods backed by Headless Services, if the user does not explicitly request a globalIP via
  GlobalnetEgressIP CRD, the same globalIP assigned to the Pods will be used for both Ingress
  as-well-as Egress. However, if user creates a GlobalnetEgressIP that matches the namespace,
  or the backend pods of Headless Service, the globalIP assigned via GlobalnetEgressIP takes

>   precedence over individual Pod globalIPs as EgressIP.

Reviewer: Isn't it expected that for headless service pods the IP that is of the pod take precedence?

It also aligns with the rest of the proposal where a specific allocation takes precedence over a broader one.

Author: > Isn't it expected that for headless service pods the IP that is of the pod take precedence?

This is to support the requirement for ExternalNetwork use-cases where we want a unique IP as EgressIP (even when Pods restart) all the time (if requested by the user).

Hence, if the user does not explicitly request a globalIP via GlobalnetEgressIP CRD, the same globalIP assigned to the Pods will be used for both Ingress as-well-as Egress. However, if user creates a GlobalnetEgressIP that matches the namespace,  or the backend pods of Headless Service, the globalIP assigned via GlobalnetEgressIP takes
precedence over individual Pod globalIPs as EgressIP.

> It also aligns with the rest of the proposal where a specific allocation takes precedence over a broader one.

Yes it still aligns. We are saying that if GlobalnetEgressIP is created, it takes the highest precedence.

CC: @mangelajo @mkimuram

Reviewer 2: That's for ingress.

For egress, if there  is an specific egress request, it's because there is a necessity for it. So it must take precedence at egress. For example, a remote firewall on a external network expecting static IPs out of the pods, or IP based authentication on a database (a'las mysql), or some sort of debugging happening with the pods over the net.

Otherwise when those pods are upgraded/rotated, the IPs will change, and such authentication / rules / debug will be broken.

The assumption is that if GlobalnetEgressIP is requested over the pods it's a higher priority on the egress.

Of course, this will not take precedence over existing traffic that has been DNATed on ingress over the pods globalnet ingress ips.

Reviewer 2: See the table in here: https://docs.google.com/document/d/1qnZ2LpF_rXGfnYYPNTldQ4WbeEUxwnuQD-xTC6GbZdg/edit#heading=h.t9ki2bstqhx0

Reviewer 4: > This is to support the requirement for ExternalNetwork use-cases where we want a unique IP as EgressIP (even when Pods restart) all the time (if requested by the user).

I agree that requirement of unique IP exists in ExternalNetwork use-cases.
However, is it only with the use cases?

I guess that it would be helpful even in cross-cluster use-cases for users who would like to have a control over IP addresses.

Author: Right. It will help even for other use-cases if the egress IP address does not change when Pods restart (f.e, network policies could be one of the use-cases)

have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
will be closer to the Vanilla Submariner experience.

## Design Details

The following CRD is proposed to support the new design.

```Go
type GlobalnetEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
    Spec               EgressIPSpec `json:"spec"`

    // Observed status of GlobalnetEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type EgressIPSpec struct {
    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
    // K8s objects, then it will be allocated and the status field will be updated accordingly.
    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
    // If this field is not specified, a single globalIP will be allocated.
    // This field cannot be changed through updates.
    // +optional

>     GlobalIPs  []string `json:"globalIPs,omitempty"`

Reviewer 2: I still believe that a list of GlobalIPs will not be very usable when the user wants to select multiple IP addresses as source but, he doesn't care about the specific IPs, that will force them to do manual allocation of those IPs.

I like the idea you proposed @sridhargaddam :+1:

```
// User can explicitly request a specific GlobalIP and this should be a
// valid IPaddress from globalnetCIDR allocated to the Cluster. 
// If the requested globalIP falls outside of the globalnetCIDR or is
// already allocated, then Status field will be set to Error with
// appropriate message.
// If this field is not specified, a single globalIP will be allocated.
// This field cannot be changed through updates.
// +optional
GlobalIP      string  `json:"globalIP,omitempty"`

// The number of globalIPs requested.
// User can specify this field while omitting GlobalIP. In such cases,
// Globalnet will allocate the requested number of contiguous GlobalIPs
// for this GlobalnetEgressIP object.
// If unspecified, NumGlobalIPs defaults to 1.
// +optional
NumGlobalIPs  int     'json:"numGlobalIPs",omitempty"`
```

    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

const (
    // GlobalnetEgressIPSuccess means that Globalnet was able to successfully
    // allocate the GlobalIP as requested in GlobalnetEgressIP object.
    GlobalnetEgressIPSuccess GlobalnetEgressIPStatus = "Success"

    // GlobalnetEgressIPError means that Globalnet was unable to allocate the
    // GlobalIP for the requested GlobalnetEgressIP object.
    GlobalnetEgressIPError GlobalnetEgressIPStatus = "Error"
)

type EgressIPStatus struct {
    // Status is one of {"Success", "Error"}
    Status GlobalnetEgressIPStatus `json:"status,omitempty"`

    // +optional
    Message *string `json:"message,omitempty"`

    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
    GlobalIPs []string `json:"globalIPs"`

    // The Namespaces to which the GlobalIPs are applied.
    Namespaces []string `json:"namespaces"`

    // The Pods to which the GlobalIPs are applied.

>     Pods []string `json:"pods"`

Reviewer: While probably easier to store this information here, I feel that if we annotate the namespaces/pods that get a global egress ip with the active policy(or policies) that affect them it'll be better for the users as they'll more easily see which policy affects their pods/namespaces.

Does this make sense?

Author: Right. I want to take some time to decide whether we want to annotate each and every Pod/Namespace as it requires an update operation. As its an implementation detail, we can decide how we want to proceed later.

If user is interested in Pod to Pod communication for all the Pods, it is recommended to install
Clusters with non-overlapping CIDRs and use Vanilla Submariner instead of Globalnet.

## Work items

1. Support creation of GlobalnetEgressIP CRD and the associated Cluster Roles.
2. Provide configuration parameter either via ConfigMap or Env variable to specify
   the number of default GlobalIPs that have to be allocated at cluster level.
3. Support Globalnet EgressIPs at the cluster level and program necessary egress rules.
4. Support Globalnet EgressIPs at namespace level which takes precedence over
   globalIPs at cluster level.
5. Support Globalnet EgressIPs to Pods (or set of Pods) which has the highest precedence.
6. Annotate only exported Services instead of annotating all Services.
7. Support Headless Services and program ingress/egress rules for the backend Pods.
8. Avoid dependency on the iptable chains programmed by IPtables kube-proxy driver.
9. Implement necessary unit tests for each of the use-cases.
10. Implement necessary e2e tests to validate various use-cases.
11. Update documentation.

> 12. Support upgrade from 0.8 release to the newer implementation.

Reviewer: How will this work? Didn't you mention the change is not backwards compatible?

Author: Done, I'll update the action item.

Inorder to support the newer implementation, a combination of ipSets with multiple IPTable chains
will be used. Also, we will require two annotations, one for ingressGlobalIP and the other for
egressGlobalIP. Because of this, backward compatibility cannot be maintained with the existing
implementation of Globalnet.

### User Stories

1.User has two (or more) clusters with Overlapping CIDRs and requires the Pods in one
  Cluster be able to discover and talk to exported Services.

  To achieve this use-case, user can simply deploy Submariner Globalnet on all the participating
  clusters. By default, a globalIP (or a set of globalIPs) will be assigned at the Cluster level
  which will be used as egressIPs for inter-cluster communication and necessary Services can be
  exported.

2.In a Globalnet deployment user wants the Pods in a namespace `ns1` to use a unique globalIP
  instead of the globalIP assigned at the cluster level.

  To achieve this, one can apply the following CRD.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: ns-egressip
   spec:
     namespaceSelector:
       matchLabels:
         project: ns1
     globalIPs:

>        - "169.254.1.30"

Reviewer 2: This bit needs update now :)

```

3.In a Globalnet deployment user wants the Pods in a namespace `ns1` to use unique globalIPs
  instead of the globalIPs assigned at the cluster level.

To achieve this, one can apply the following CRD.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: ns-egressip
   spec:
     namespaceSelector:
        matchLabels:
          project: ns1
     globalIPs:
       - "169.254.1.30"

>        - "169.254.1.32"

Reviewer 2: Needs update now

remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
will be closer to the Vanilla Submariner experience.

## Design Details

The following CRD is proposed to support the new design.

```Go
type ClusterGlobalEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of ClusterGlobalEgressIP object.
    Spec               ClusterGlobalEgressIPSpec `json:"spec"`

    // Observed status of ClusterGlobalEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type ClusterGlobalEgressIPSpec struct {
    // The number of globalIP's requested at the cluster-level.
    // Globalnet Controller will allocate the requested number of contiguous GlobalIPs for
    // this ClusterGlobalEgressIP object.
    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
    // +optional

>     NumGlobalIPs  *int     'json:"numGlobalIPs",omitempty"`

Reviewer 2: just a nit, we are shuttering with Global/Global. May be we should just use Size or NumIPs?

But a nit..

Author: I'm fine with it. We just merged the corresponding CRD in submariner, but it should not be a problem to update it.
@tpantelis what are your thoughts?

Reviewer 5: NumIPs sounds fine.  Size seem a bit vague.

}

type GlobalEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalEgressIP object.
    Spec               GlobalEgressIPSpec `json:"spec"`

    // Observed status of GlobalEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type GlobalEgressIPSpec struct {
    // The number of globalIP's requested at the namespace level or for selected pods in a namespace.
    // Globalnet Controller will allocate the requested number of contiguous GlobalIPs for this
    // GlobalEgressIP object.
    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
    // +optional

>     NumGlobalIPs  *int     'json:"numGlobalIPs",omitempty"`

Reviewer 2: Size or NumIPs ?

Author: ditto :) 
@tpantelis

have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
will be closer to the Vanilla Submariner experience.

## Design Details

The following CRD is proposed to support the new design.

```Go
type GlobalnetEgressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
    Spec               EgressIPSpec `json:"spec"`

    // Observed status of GlobalnetEgressIP. Its a read-only field.
    // +optional
    Status             EgressIPStatus `json:"status,omitempty"`
}

type EgressIPSpec struct {
    // User can explicitly request a specific GlobalIP and this should be a valid IPaddress
    // from globalnetCIDR allocated to the Cluster. If the requested globalIP falls outside
    // of the globalnetCIDR or is already allocated (either fully or partially), then Status
    // field will be set to Error with appropriate message.
    // If this field is not specified, a single globalIP will be allocated.
    // This field cannot be changed through updates.
    // +optional

>     GlobalIP      string  `json:"globalIP,omitempty"`

Reviewer: After thinking about it some more, how much sense does it really make to allow to specify an explicit IP address here?

In K8s AFAIK, except very specific cases of using LoadBalancer Services on specific clouds, you can't specify an IP address. K8s takes care of IP address allocations.

In ingress global IP use case, the globalnet takes care of IP allocation.

Wouldn't it make more sense for the system to allocate the egress IP as well?

For firewall considerations the users can retrieve the IP once allocated and use that to define the necessary firewall rules.

For a system perspective, allowing to specify the IP seems like a more risky endeavour than having globalnet control allocation (e.g. it can more effectively prevent race conditions, for example).

Reviewer 4: Specifying GlobalIPs provides ability to:
1. Keep the GlobalIP unchanged,
2. Specify the same GlobalIP to certain resources.

If these abilities are provided by the other way, directly specifying the globalIPs might not be necessary (I'm not sure that I created an exclusive list, so someone might find other requirements, later).

Note that I assume that users would expect the globalIP to be unchanged beyond the lifecycle of the namespace or resources inside it. And it reminds me of abstractions like SC/PVC/PV and Elastic IP, which might be too complex for users in this case.

Author: @mkolesnik agree with you that K8s automatically allocates an IPaddress to the Pod. But we also have some CNIs (for example - [calico](https://docs.projectcalico.org/networking/use-specific-ip)) that allow specifying a PodIP. In general, for most of the use-cases, users may not be specifying the GlobalIP, but it would be helpful sometimes as @mkimuram mentioned. 

For example, let's say user deployed a Pod and it is assigned some globalIP. User then configures the external-firewall/application/network-policies to use the globalIP. If the Pod gets restarted, unless there is a provision to specify GlobalIP, a different globalIP could be assigned to the Pod.

Reviewer: > Specifying GlobalIPs provides ability to:
>     1. Keep the GlobalIP unchanged,

It is expected in this proposal that the GlobalIP(s) assigned to a specific CR will indeed be consistent and not suddenly change.

> 
>     2. Specify the same GlobalIP to certain resources.

Indeed that's what the proposal indicates, I'm not suggesting we change that.

> 
> 
> If these abilities are provided by the other way, directly specifying the globalIPs might not be necessary (I'm not sure that I created an exclusive list, so someone might find other requirements, later).

Indeed it will be provided either way.

> 
> Note that I assume that users would expect the globalIP to be unchanged beyond the lifecycle of the namespace or resources inside it. And it reminds me of abstractions like SC/PVC/PV and Elastic IP, which might be too complex for users in this case.

> @mkolesnik agree with you that K8s automatically allocates an IPaddress to the Pod. But we also have some CNIs (for example - [calico](https://docs.projectcalico.org/networking/use-specific-ip)) that allow specifying a PodIP. In general, for most of the use-cases, users may not be specifying the GlobalIP, but it would be helpful sometimes as @mkimuram mentioned.

It seems though that he's fine with not having the ability to specify a specific IP, as long as the IP stays consistent (which it should).

> 
> For example, let's say user deployed a Pod and it is assigned some globalIP. User then configures the external-firewall/application/network-policies to use the globalIP. If the Pod gets restarted, unless there is a provision to specify GlobalIP, a different globalIP could be assigned to the Pod.

Why would that happen? I merely suggested that globalnet controller be allocating the IPs, without allowing the user to do so.

We could always add the ability to specify a specific IP in a later stage if thats really a valuable feature, but to me it seems unnecessary for now.

Author: > It seems though that he's fine with not having the ability to specify a specific IP, as long as the IP stays consistent (which it should).

Indeed, the globalIP assigned to a Pod/Service will remain consistent through the life-cycle of the resource. @mkimuram can you please confirm if it's okay to remove the ability to specify globalIP as part of the `GlobalnetEgressIP` request?

> > For example, let's say user deployed a Pod and it is assigned some globalIP. User then configures the external-firewall/application/network-policies to use the globalIP. If the Pod gets restarted, unless there is a provision to specify GlobalIP, a different globalIP could be assigned to the Pod.
> 
> Why would that happen? I merely suggested that globalnet controller be allocating the IPs, without allowing the user to do so.

Yeah, when the user creates a `GlobalnetEgressIP` object, a globalIP will be allocated to the matching Pods. As long as the object is not deleted, the same globalIP will be re-used even when the Pods are restarted. 

> We could always add the ability to specify a specific IP in a later stage if thats really a valuable feature, but to me it seems unnecessary for now.

I'm fine with this, if its Okay to others. @mangelajo @mkimuram @aswinsuryan @vthapar

Reviewer 4: @sridhargaddam 

> Indeed, the globalIP assigned to a Pod/Service will remain consistent through the 
> life- cycle of the resource. @mkimuram can you please confirm if it's okay to remove 
> the ability to specify globalIP as part of the GlobalnetEgressIP request?

As commented in the meeting, I agree with it for initial implantation. Let's consider adding it later, if we actually get feedback from users that it is really needed.

    // in case it is not set, it results in all the Pods selected from the NamespaceSelector.
    // In case it is set, its intersected with NamespaceSelector, and only Pods that match the
    // PodSelector and present in the NamespaceSelector will have the GlobalIPs as EgressIPs.
    // Either namespaceSelector or podSelector have to be specified. If both are omitted,
    // its considered as an error.
    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

const (
    // GlobalnetEgressIPSuccess means that Globalnet was able to successfully
    // allocate the GlobalIP as requested in GlobalnetEgressIP object.
    GlobalnetEgressIPSuccess GlobalnetEgressIPStatus = "Success"

    // GlobalnetEgressIPError means that Globalnet was unable to allocate the
    // GlobalIP for the requested GlobalnetEgressIP object.
    GlobalnetEgressIPError GlobalnetEgressIPStatus = "Error"
)

type EgressIPStatus struct {
    // Status is one of {"Success", "Error"}

>     Status GlobalnetEgressIPStatus `json:"status,omitempty"`

Reviewer 2: I think we should move this to a list of conditions, like it's standard in k8s. So the admin can see the progress of the object. 

May be there will be not much progress in this type of object though, what do you think?

Author: I'm not able to think of any progress for this type except for `Success` or `Error`. Hence did not use the `Conditions`

Reviewer: Yes probably the Success/Error could be a condition since it fits the bill, other fields not so much.

    // +optional
    Message *string `json:"message,omitempty"`

    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
    GlobalIPs []string `json:"globalIPs"`

    // The Namespaces to which the GlobalIPs are applied.
    Namespaces []string `json:"namespaces"`

    // The Pods to which the GlobalIPs are applied.
    Pods []string `json:"pods"`
}
```

When Submariner Globalnet is deployed in a Cluster, it auto-allocates the configured number of
`globalIPs` and programs the necessary egress rules on the active Gateway node of the Cluster.
This is done internally by creating GlobalnetEgressIP CR object with the name `cluster-default`.
Users can query the CR to check the `Status.globalIPs` allocated at the Cluster level. If users
create multiple GlobalnetEgressIP CRs that match the Cluster scope, one of the egressIP at the
Cluster scope will be used.

Creation of GlobalnetEgressIP object will be controlled via a ClusterRole and only users having
the necessary role will be able to create the objects. The GlobalnetEgressIP object can be

> created in any namespace with selectors that match any other namespaces.

Reviewer 4: @sridhargaddam 

Do we need to be clear who will have the ClusterRole?

I assume that there will be cluster admin and users for a particular namespace, at least.

Cluster admin obviously should be able to configure all of them.
On the other hand, users for a particular namespace isn't expected to be able to configure the one for other namespace in most cases.

Can we make it controllable or should we just don't allow latter use case? 
(Or do we allow users who have the ClusterRole to have control over all namespaces, as the current model does.)

Reviewer 6: @mkimuram Only admin will have access to create these objects right?  unless a user is granted access for a specific reason,  going by the principle of least privilege.

Reviewer 4: I think that some users will find it useful if they can configure IP addresses of the resources in their namespaces. And they basically won't care for other namespaces. If we can allow them to configure it only for their namespace, it should satisfy the principle of least privilege. 

If we only allow admin to configure it, which is not what I expect, we should consider making this CRD cluster scoped resource or allowing it to exist only in a specific namespace, like submariner-operator namespace. It would make it easier to check what is configured.  

Then, should we consider separating this CRD to one for admin to configure cluster level and the other for user to configure namespace level?

Reviewer: It does sound weird that in a certain namespace there could be a CR imposing policy on other namespaces.

Perhaps a more coherent model would be that a CR on a namespace would only affect pods in that namespace, and if no pod selector is specified then it affects all pods in the namespace?
Then namespace selector is implicit and doesn't need to exist as a field.

Cluster wide settings can then be applied on the `submariner` namespace in a similar fashion allowing either all pods or select pods.

WDYT?

Author: We discussed this point in today's external network meeting and agreed on the following. @mkolesnik/@aswinsuryan  if you have any comments on this, do let us know.

The existing `GlobalnetEgressIP` CRD will be modified to Cluster scope and only admin will have the provision to create the objects that apply to various namespaces. 
In the future, based on the feedback, we can introduce a different CRD that does NOT include the NamespaceSelectors and will select pods only in the namespace where the object is created. Such objects can be created by users only in their respective namespace based on the clusterRole. This will be less confusing IMHO.

Reviewer: Did you discuss https://github.com/submariner-io/enhancements/pull/22/files#r587478495 ?

IMHO it's much more intuitive and a better UX then what is proposed, BUT if it's already been discussed and rejected then we could go with the agreed upon solution (cluster level CRs).

I think the main point here is the UX and RBAC - should a user who's not an admin be able to define a globalnet policy for his namespace/pods?
It seems reasonable to me not to have to rely on admin role for this, as the cloud & k8s models are much about self service and not relying on request fulfilment from 3rd party (ie some admin or another).

To simplify we could have cluster level policy not set by this CRD at all since it's really a singleton policy for the cluster, and the CRD could be namespace scoped and define a policy for pods in that namespace.

On the UX I imagine this would be abstracted from the user anyway by subctl, so let's get the right design that makes most sense.

Author: > It seems reasonable to me not to have to rely on admin role for this, as the cloud & k8s models are much about self service and not relying on request fulfilment from 3rd party (ie some admin or another).

Agree, it makes sense and it aligns more with K8s models. 

```
Perhaps a more coherent model would be that a CR on a namespace would only affect pods in that namespace, and if no pod selector is specified then it affects all pods in the namespace?
Then namespace selector is implicit and doesn't need to exist as a field.

Cluster wide settings can then be applied on the submariner namespace in a similar fashion allowing either all pods or select pods.
```
Taking a re-look at your suggestion, if I understood it correctly, you are suggesting that if admin creates the object in `submariner-operator` namespace then we treat it as `cluster-wide`. Is this correct @mkolesnik ?

Assuming yes, what I see missing is that we do not have a provision to select specific `namespaces` (as we will be removing `Spec.NamespaceSelector`) and the CR/object will either affect the local namespace or is cluster-wide. Personally, this looks okay to me and we can add this support in a later release if required. CC: @mangelajo @mkimuram @vthapar @aswinsuryan 
 
> To simplify we could have cluster level policy not set by this CRD at all since it's really a singleton policy for the cluster, and the CRD could be namespace scoped and define a policy for pods in that namespace.

I did not understand this completely. Will try to sync up with you over slack.

Author: I discussed this with @mkolesnik and @vthapar over slack and we felt that it's better to create a separate CR that is cluster-wide and name it as `ClusterGlobalEgressIP`. 

At the same time rename the current CR to `NamespaceGlobalEgressIP` 

Only admins will have the provision to create the `ClusterGlobalEgressIP` CR and regular users will only be able to create `NamespaceGlobalEgressIP` that applies to their own namespaces.

will be used. Also, we will require two annotations, one for ingressGlobalIP and the other for
egressGlobalIP. Because of this, backward compatibility cannot be maintained with the existing
implementation of Globalnet.

### User Stories

1.User has two (or more) clusters with Overlapping CIDRs and requires the Pods in one
  Cluster be able to discover and talk to exported Services.

  To achieve this use-case, user can simply deploy Submariner Globalnet on all the participating
  clusters. By default, a globalIP (or a set of globalIPs) will be assigned at the Cluster level
  which will be used as egressIPs for inter-cluster communication and necessary Services can be
  exported.

2.In a Globalnet deployment user wants the Pods in a namespace `ns1` to use a unique globalIP
  instead of the globalIP assigned at the cluster level.

  To achieve this, one can apply the following CRD.

```yaml
   apiVersion: submariner.io/v1alpha1
   kind: GlobalnetEgressIP
   metadata:
     name: ns-egressip
   spec:
     namespaceSelector:
       matchLabels:
         project: ns1
     globalIP: "169.254.1.30"

>      numGlobalIPs: 1

Reviewer: Wouldn't the default be `1`?

Otherwise, I feel this scenario is the same as the next one.

Author: yes, the default would be 1 but there will not be any harm even if its explicitly specified. In this use-case we are showing how to request a single globalIP whereas in the next one a set of globalIPs.

    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

const (
    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"

    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
    GlobalEgressIPError GlobalEgressIPStatus = "Error"
)

type EgressIPStatus struct {
    // Status is one of {"Success", "Error"}
    Status GlobalEgressIPStatus `json:"status,omitempty"`

    // +optional
    Message *string `json:"message,omitempty"`

    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
    GlobalIPs []string `json:"globalIPs"`
}

type GlobalIngressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalIngressIP object.
    Spec GlobalIngressIPSpec `json:"spec"`

    // Observed status of GlobalIngressIP. Its a read-only field.
    // +optional
    Status GlobalIngressIPStatus `json:"status,omitempty"`
}

type GlobalIngressIPSpec struct {
    // TypeOfObject can be ClusterIPService or a Pod

>     TypeOfObject string `json:"typeOfObject"`

Reviewer 5: ```suggestion
    Target TargetType `json:"target"`
```

    // If the TypeOfObject points to a Pod that belongs to a Headless Service, it uses the
    // name of the corresponding Service.
    // +Optional
    ServiceName string `json:"serviceName,omitempty"`
}

type GlobalIngressIPStatus struct {
    // +optional
    Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`

    // +optional

>     Message *string `json:"message,omitempty"`

Reviewer 5: Not needed - this is part of `metav1.Condition`.

```suggestion
```

    // The GlobalIP assigned to the respective object.
    // +optional

>     GlobalIPs []string `json:"globalIPs"`

Reviewer 5: ```suggestion
    GlobalIP string `json:"globalIP"`
```

}

> type TypeOfObject string
> 
> const (
>     ClusterIPService TypeOfObject = "ClusterIP"
>     Pod              TypeOfObject = "Pod"
> )

Reviewer 5: ```suggestion
type TargetType string

const (
    ClusterIPService TargetType = "ClusterIPService"
    Pod              TargetType = "Pod"
)
```

> const (
>     GlobalIngressIPAllocated GlobalIngressIPConditionType = "Allocated"
>     GlobalIngressIPError     GlobalIngressIPConditionType = "Error"
> )

Reviewer 5: Not needed here.

```suggestion
```

    // Globalnet Controller will allocate the requested number of contiguous GlobalIPs for this
    // GlobalEgressIP object.
    // If unspecified, NumGlobalIPs defaults to 1 and max allowed is restricted to 10.
    // +optional
    NumGlobalIPs  *int     'json:"numGlobalIPs",omitempty"`

    // Selects the pods whose label match the definition. This is an optional field and
    // in case it is not set, it results in all the Pods selected from the namespace in which the
    // CR is created.
    // +optional
    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
}

const (
    // GlobalEgressIPSuccess means that Globalnet was able to successfully allocate the GlobalIP.
    GlobalEgressIPSuccess GlobalEgressIPStatus = "Success"

    // GlobalEgressIPError means that Globalnet was unable to allocate the GlobalIP.
    GlobalEgressIPError GlobalEgressIPStatus = "Error"
)

type EgressIPStatus struct {
    // Status is one of {"Success", "Error"}
    Status GlobalEgressIPStatus `json:"status,omitempty"`

    // +optional
    Message *string `json:"message,omitempty"`

    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.

>     GlobalIPs []string `json:"globalIPs"`

Reviewer 5: How about `AllocatedIPs`?

Author: At lot of places in this PR as well as in our doc we use globalIP or globalIPs. But if others prefer `AllocatedIPs`, I can change it.

Reviewer 5: We can still use the term _globalIP_ - `AllocatedIPs` would refer to the globalIPs allocated for this instance. I only mentioned this b/c we decided to drop _Global_ from `NumIPs` so perhaps we should do so here for consistency.

Author: Its fine to me. Lets hear from @mangelajo @mkimuram

Reviewer 4: As for the consistency, `GlobalEgressIP` now contains `GlobalEgressIPSpec` and `EgressIPStatus`, it sounds a bit inconsistent. 

I don't have a strong preference on removing `Global` from everywhere, but it would be possible to just 's/Global//', except `GlobalIngressIPStatus.GlobalIP` (it should be something like `IngressIPStatus.AllocatedIP`), without conflict and significant meaning change (unless there are non-global EgressIP or IngressIP object for other purpose).

The other direction is just move `EgressIPStatus` back to `EgressGlobalIPStatus`, but I prefer above way. In both case, I agree with changing to `AllocatedIP`(s).

Reviewer 5: In https://github.com/submariner-io/submariner/pull/1172, I named it `GlobalEgressIPStatus` for consistency. I thought about dropping the `Global` prefix from the structs but doc comments refer to "_GlobalIPs_" and "_Globalnet_" so it seems they're implicitly tied to Globalnet.

Author: Thank you @mkimuram @tpantelis. I'll use AllocatedIPs

Reviewer 5: Ok I pushed https://github.com/submariner-io/submariner/pull/1334 to update the types.  I'd like to suggest that we continue reviewing the type implementations in that PR. I think this PR should be primarily for concepts plus this one has gotten quite large 😄

    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
    GlobalIPs []string `json:"globalIPs"`
}

type GlobalIngressIP struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec is the specification of desired behavior of GlobalIngressIP object.
    Spec GlobalIngressIPSpec `json:"spec"`

    // Observed status of GlobalIngressIP. Its a read-only field.
    // +optional
    Status GlobalIngressIPStatus `json:"status,omitempty"`
}

type GlobalIngressIPSpec struct {
    // TypeOfObject can be ClusterIPService or a Pod
    TypeOfObject string `json:"typeOfObject"`

    // If the TypeOfObject points to a Pod that belongs to a Headless Service, it uses the
    // name of the corresponding Service.
    // +Optional
    ServiceName string `json:"serviceName,omitempty"`

> }

Reviewer 4: How will the relationship between `GlobalIngressIP` and `Service`/`Pod` be stored?
If it relies on that namespace and name are the same between resources, I'm worrying about the conflict on name collision of `Service` and `Pod`. The same resource with the same namespace and name can't be created, while different resource with the same namespace and name can be created. As a result, if `Service` and `Pod` have the same namespace and name, creation of `GlobalIngressIP` will fail.

Simple example is:
```
kubectl create -f https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/application/web/web.yaml
kubectl run nginx --image=nginx
kubectl expose pod nginx --port=80 --name=web-0

kubectl get all
NAME        READY   STATUS    RESTARTS   AGE
pod/nginx   1/1     Running   0          33s
pod/web-0   1/1     Running   0          4m51s
pod/web-1   1/1     Running   0          4m47s

NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   10.0.0.1     <none>        443/TCP   3h25m
service/nginx        ClusterIP   None         <none>        80/TCP    4m52s
service/web-0        ClusterIP   10.0.0.129   <none>        80/TCP    4s

NAME                   READY   AGE
statefulset.apps/web   2/2     4m52s
```

Reviewer 3: @mkimuram In such a scenario, why would one need IngressIP for that Pod? As you mentioned in earlier comment, analogous for IngressIP is Endpoints. In scenario you shared, how does k8s handle naming conflict for Endpoints? If the `pod/nginx` doesn't have an Endpoint, why would it need an IngressIP?

```
kubectl get endpoints
NAME         ENDPOINTS                   AGE
kubernetes   172.31.0.5:6443             8m47s
nginx        10.1.0.3:80,10.1.128.1:80   2m40s
web-0        10.1.0.2:80                 2m27s
```
In this case the IngressIP will be for the `Service` not the pod. I discussed with Sridhar that for implementation Globalnet will use Endpoint events instead of Pod to know when to assign IngressIP for it. Idea being, if pod is not reachable from within cluster [no endpoint] then we don't want to allocate GlobalIP to it either.

Makes sense?

EDIT:
Discussed with Sridhar and he actually brought up a better example of conflict. He'll share his comments later.

Reviewer 2: It is a good question,

Headless services would need an GlobalIngresIP I believe @sridhargaddam can confirm

Could we append a postfix to auto-generated resources, for example web-0-svc GlobalIngressIP, or web-0-pod ? 

We could also apply a randomization pattern postfix to pod generated GlobalIngressIP only (as k8s does to deployment pods )

Author: That's a good question @mkimuram. 

Yes @mangelajo, currently the GlobalIngressIP object will be created for ClusterIP Service (only one object) and for every Backend Pod in a Headless Service backed by Deployment or StatefulSets.

@mkimuram, as you pointed out, the issue could be seen while Globalnet tries to create a GlobalIngressIP object if we just use the name of Service or Pod alone. I discussed with @vthapar offline and we felt that we can prepend `svc-` or `pod-` while naming the `GlobalIngressIP` objects.

i.e., for the example where we have a ClusterIPService named `web-0` and HeadlessService with two backendPods named `web-0` and `web-1`
```
kubectl get globalIngressIPs -n default
svc-web-0
pod-web-0
pod-web-1
```

Along with this, to avoid ambiguity, I'll update the constants (which is an internal field) as shown below. 
```
const (
    ClusterIPService         TargetType = "ClusterIPService"
    HeadlessServicePod   TargetType = "HeadlessServicePod"
)
```

Note: GlobalIngressIP.Spec.ServiceName will contain the name of the corresponding Service.

Is this Okay @mkimuram, @mangelajo ?

Reviewer 4: Thank you for sharing the idea.

I agree that it will work in most cases. There will be one corner case where length of the name + prefix exceeds the [maximum length](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names). In this case, the name can still collide.

I guess that we can go ahead with the idea by just adding additional fallback logic (something like, truncate name for long one, add suffix, match with patterns, and check if it is really for the resource). For that purpose, I think that `v1.LocalObjectReference` should be added to `GlobalIngressIPSpec` (Just adding name as string will also work, for we already have `TargetType`. Also, `v1.TypedLocalObjectRef` or `v1.ObjectReference` may be too much for this purpose).

I think that this fallback implementation can be TODO with name length restriction until it will be used in production, but it would be better to add the field to API with this case in mind.

Author: Done, thanks @mkimuram
